---
title: "Microbiota MSP Notebook"
output: html_notebook
---

## Load library

```{r, message=FALSE, warning=FALSE}

library(dplyr)
library(magrittr)
library(ggplot2)
library(vegan)

devtools::load_all()


library(IBSMicrobiota)

```

## MSP wrangling

the aim here is to summarize the count data to MSP

```{r}

data("hs3_9.genes_length")
data("counts")
data("MSP_tax")
data("MSP_genes_id_corresp")



counts %>% head



```

### Normalize by gene length and read count

```{r}

merge(counts, hs3_9.genes_length, by.x="id_fragment", by.y="row.names") %>%
  select(-id_fragment, -id_fragment_external, -sample_count) -> counts


cols = grep("M", colnames(counts), value=TRUE)

total_reads = apply(counts[cols],2,sum)

remove_sample = which(total_reads < 1000000) %>% names



counts[cols] = lapply(counts[cols], `/`, counts$length) 


#


#counts_rescaled = counts %>% select(-one_of(remove_sample))

counts[cols] = prop.table(counts[cols] %>% as.matrix ,2)

counts_rescaled = counts

counts_rescaled[cols] = counts_rescaled[cols] %>% t %>% multiply_by(total_reads %>% as.numeric ) %>% t %>% round()


counts = counts %>% select(-one_of(remove_sample))

#cols_remove_sample = cols[-remove_sample]

#which(cols == "M066")

#which(counts %>% colnames == "M066")

```

### export gene richness

```{r}

#total_reads %>% sort

genes_richness_1M =

counts_rescaled %>%
  select(-M077, -M004, -M073, -M050, -M096, -M098, -genes.id, -length) %>%
  t %>%
  vegan::rarefy(sample = 10^6)

genes_richness_1M %>%
  as.matrix() %>%
  as.data.frame %>%
  dplyr::rename(richness=V1) %>%
  write.csv2(file="genes_richness.csv")
  
  
genes_richness_1M %>%
  as.matrix() %>%
  as.data.frame %>%
  dplyr::rename(richness=V1) %>%
  ggplot() + geom_density(aes(x=richness),col="black")




```




### make the correspondance with MSP

```{r}

merge(counts, MSP_genes_id_corresp %>% select(msp_name,module_name,gene_name_v2), by.x="genes.id", by.y="gene_name_v2") %>%
  filter(module_name == "core") %>%
  group_by(msp_name) %>%
  #select(-module_name, -length, -genes.id, cols) %>%
  #.[,cols_remove_sample] %>%
  select(-length,-module_name,-genes.id) %>%
  summarize_all(funs(median)) -> MSP_abundance

MSP_abundance %>% head

write.csv2(MSP_abundance, file="MSP_abundance.csv")





```

### MSP read mass

```{r}

merge(counts_rescaled, MSP_genes_id_corresp %>% select(msp_name,module_name,gene_name_v2), by.x="genes.id", by.y="gene_name_v2") %>%
  group_by(msp_name) %>%
  select(-length,-module_name,-genes.id) %>%
  summarize_all(funs(sum)) -> MSP_read_mass


MSP_read_mass %>% 
  reshape2::melt(id.vars="msp_name") %>% 
  group_by(variable) %>% 
  summarize(value=sum(value)) %>%
  merge(., total_reads %>% as.matrix %>% as.data.frame, by.x="variable", by="row.names") %>%
  mutate(MSP_mass = value/V1) %>%
  ggplot() + geom_density(aes(x=MSP_mass)) + xlim(0,1)




```



### MSP prevalence


```{r  , include=FALSE}

prevalent_MSP = 
MSP_abundance %>% 
  reshape2::melt(id.vars="msp_name") %>%
  group_by(msp_name) %>%
  filter(value > 0) %>%
  summarize(n=n()) %>%
  mutate(p = n/(dim(MSP_abundance)-1)) %>%
  filter(p>0.25) %>%
  merge(.,MSP_tax, by.y = "MSP_ID", by.x = "msp_name") %>%
  arrange(n)



```



### correspondance between MSP and taxonomy
```{r}

dim(MSP_abundance)

MSP_tax %>% head

merge(MSP_abundance, MSP_tax, by.x="msp_name", by.y="MSP_ID") %>%
  group_by(species,genus,family,order,class,phylum,superkingdom) %>%
  select(-msp_name) %>%
  summarize_all(funs(sum)) -> MSP_species_abund

MSP_species_abund %>% dim

merge(MSP_abundance, MSP_tax, by.x="msp_name", by.y="MSP_ID") %>%
  group_by(genus,family,order,class,phylum,superkingdom) %>%
  select(-msp_name,-species) %>%
  summarize_all(funs(sum)) -> MSP_genus_abund

MSP_genus_abund %>% dim

MSP_species_abund[cols[!(cols %in% remove_sample)]] = prop.table(MSP_species_abund[cols[!(cols %in% remove_sample)]] %>% as.matrix,2)
MSP_genus_abund[cols[!(cols %in% remove_sample)]]   = prop.table(MSP_genus_abund[cols[!(cols %in% remove_sample)]]   %>% as.matrix,2)

cols_removed = cols[!(cols %in% remove_sample)]

```



## Check some species distribution


```{r, fig.height=8, fig.width=10, message=FALSE, warning=FALSE}



 
MSP_species_abund %>%
  filter(genus == "Prevotella") %>%
  reshape2::melt(id.vars=c("species","genus","family","order","class","phylum","superkingdom")) %>%
  ggplot() + geom_density(aes(x=value+10^-6)) + facet_wrap(~species, scale="free_y") + scale_x_log10()
  
MSP_species_abund %>%
  filter(genus == "Methanobrevibacter") %>%
  reshape2::melt(id.vars=c("species","genus","family","order","class","phylum","superkingdom")) %>%
  ggplot() + geom_density(aes(x=value+10^-6)) + facet_wrap(~species, scale="free_y") + scale_x_log10()
  
  
MSP_species_abund %>%
  filter(genus == "Ruminococcus") %>%
  reshape2::melt(id.vars=c("species","genus","family","order","class","phylum","superkingdom")) %>%
  ggplot() + geom_density(aes(x=value+10^-6)) + facet_wrap(~species, scale="free_y") + scale_x_log10()
  
MSP_species_abund %>%
  filter(genus == "Faecalibacterium") %>%
  reshape2::melt(id.vars=c("species","genus","family","order","class","phylum","superkingdom")) %>%
  ggplot() + geom_density(aes(x=value+10^-6)) + facet_wrap(~species, scale="free_y") + scale_x_log10()
  
  
MSP_species_abund %>%
  filter(genus == "Bacteroides") %>%
  reshape2::melt(id.vars=c("species","genus","family","order","class","phylum","superkingdom")) %>%
  ggplot() + geom_density(aes(x=value+10^-6)) + facet_wrap(~species, scale="free_y") + scale_x_log10()
   

```


check associations between MSP and 16S data

```{r}

#devtools::load_all("../../IBSMicrobiota")
data("IBSData")

apply(IBSData$otu[, 


IBSData$metadata %>%
  select(Patient_ID,Visit,Sample_type,Sample_ID) %>%
  filter(Visit == "V4", Sample_type == "Stool") %>%
  filter(Patient_ID %in% cols_removed) %>% .$Sample_ID], 2, 
tapply, paste(IBSData$tax$phylum,IBSData$tax$family,IBSData$tax$genus,sep="_"), sum) %>%
  prop.table(2) %>% as.data.frame -> genus16S
  
  colnames(genus16S) %<>% stringr::str_sub(.,start=0, end=4)
#genus16S[cols_removed]

cor(MSP_genus_abund[-c(1:6)] %>% t, genus16S[cols_removed] %>% t, method="spearman") %>% 
  reshape2::melt() %>% na.omit() %>% 
  filter(abs(value) > 0.65) %>% cbind(MSP_genus_abund[.$Var1,c(2)],MSP_genus_abund[.$Var1,c(1)],. ) %>% 
  arrange(desc(value))



```

## Check raw gene abundance within pangenome




```{r, eval=FALSE, include=FALSE}


MSP_tax %>%
  filter(genus == "Prevotella")


merge(counts_rescaled, MSP_genes_id_corresp %>% select(msp_name,module_name,gene_name_v2), by.x="genes.id", by.y="gene_name_v2") %>%
  filter(msp_name=="msp_0039") -> msp_0039_counts

merge(counts_rescaled, MSP_genes_id_corresp %>% select(msp_name,module_name,gene_name_v2), by.x="genes.id", by.y="gene_name_v2") %>%
  filter(msp_name=="msp_0560") -> msp_0560_counts


merge(counts_rescaled, MSP_genes_id_corresp %>% select(msp_name,module_name,gene_name_v2), by.x="genes.id", by.y="gene_name_v2") %>%
  filter(msp_name=="msp_0023") -> msp_0023_counts


msp_0039_counts %>% head

msp_0039_counts %>%
  select(-length) %>%
  reshape2::melt(id.vars=c("genes.id","msp_name","module_name")) %>%
  group_by(variable) %>%
  #filter(value != 0) %>%
  mutate(value_scaled=scale(value),med=median(value), module_name_2 = ifelse(module_name == "core","core","other")) %>%
  filter(variable %in% c("M056")) %>%
  arrange(value) %>%
  mutate(genes.id = factor(genes.id , levels = genes.id)) %>%
  ggplot() + geom_point(aes(x=genes.id,y=value), stat="identity", size=0.1) + 
  #geom_point(aes(x=genes.id,y=med), stat="identity", size=1, col="red") +
  facet_grid(variable~module_name_2, scale="free_x") +
  theme(axis.text.x=element_blank(), axis.ticks.x = element_blank())


rbind(msp_0560_counts,
msp_0039_counts,msp_0023_counts) %>%
  select(-length) %>%
  reshape2::melt(id.vars=c("genes.id","msp_name","module_name")) %>%
  group_by(variable,msp_name) %>%
  filter(module_name=="core") %>%
  summarise(value_min=quantile(value,0.025), value_max=quantile(value,0.975),value_med=median(value)) %>%
  ggplot() + geom_point(aes(x=value_min+1,y=value_med+1)) + scale_y_log10() + scale_x_log10() + facet_wrap(~msp_name)

msp_0560_counts %>%
  select(-length) %>%
  reshape2::melt(id.vars=c("genes.id","msp_name","module_name")) %>%
  group_by(variable) %>%
  #filter(value != 0) %>%
  mutate(value_scaled=scale(value),med=median(value), module_name_2 = ifelse(module_name == "core","core","other")) %>%
  filter(variable %in% c("M013")) %>%
  arrange(value) %>%
  mutate(genes.id = factor(genes.id , levels = genes.id)) %>%
  ggplot() + geom_point(aes(x=genes.id,y=value), stat="identity", size=0.1) + 
  #geom_point(aes(x=genes.id,y=med), stat="identity", size=1, col="red") +
  facet_grid(variable~module_name_2, scale="free_x") +
  theme(axis.text.x=element_blank(), axis.ticks.x = element_blank())


```



## from binary gene data to infraspecies

```{r}


MSP_tax %>%
  filter(genus == "Bacteroides") #check thetaiotamicron



merge(counts_rescaled, 
      MSP_genes_id_corresp %>% 
        select(msp_name,module_name,gene_name_v2) %>% 
        filter(msp_name %in% c("msp_0003")),
      by.x="genes.id", by.y="gene_name_v2") -> msp_counts



# calculate mapping read statistics on core msp and select only sample with an average of 2 reads per gene 


msp_counts %>%
  select(-length) %>%
  reshape2::melt(id.vars=c("genes.id","msp_name","module_name")) %>%
  group_by(variable) %>%
  filter(module_name=="core") %>%
  summarise(value_min=quantile(value,0.025), value_max=quantile(value,0.975),value_med=median(value)) %>%
  filter(value_med >= 2) -> raw_reads_core_stat


# convert abundance data to binary matrix

msp_counts %>%
  select(-length) %>%
  reshape2::melt(id.vars=c("genes.id","msp_name","module_name")) %>%
  merge(.,raw_reads_core_stat, by=c("variable"), all_y = TRUE) %>%
  mutate(bin = ifelse(value<value_max,ifelse(value>value_min+1,1,0),0)) %>%
  reshape2::dcast(genes.id+module_name~variable, value.var="bin") %>%
  mutate(module_name = ifelse(module_name=="core","core","other")) -> binary_msp


heatmap(binary_msp %>% filter(module_name  != "core") %>% select(-module_name,-genes.id) %>% as.matrix, 
        scale = "none",cexRow = 0.01, 
        #RowSideColors = c("blue","red")[binary_msp %>% filter(module_name  != "core") %>% .$module_name %>% as.factor %>% as.numeric],  
        col=c("black","yellow"))

# compute clusters from binary matrix using non MSP core genes

binary_msp %>% filter(module_name  != "core") %>% 
  select(-module_name,-genes.id) %>% as.matrix %>%
  t %>%
  ade4::dist.binary(.,method=1) %>%
  fpc::clusterboot(.,B=100,bootmethod=
                                 "subset",clustermethod=fpc::pamkCBI, count=FALSE,
                               k=1:6, showplot=FALSE) -> msp_clust

msp_clust$partition

# plot heatmap of MSP clusters (Infraspecies)


heatmap(binary_msp %>% filter(module_name  != "core") %>% select(-module_name,-genes.id) %>% as.matrix, 
        scale = "none",cexRow = 0.01, 
        ColSideColors = c("blue","red","green","gray")[msp_clust$partition %>% as.factor %>% as.numeric],  
        col=c("black","yellow"))


```

## use inbuilt function

```{r}




MSP_tax %>%
  filter(genus == "Bacteroides") #check stecoris


MSP_tax %>%
  filter(species == "Eubacterium rectale") #check rectale




merge(counts_rescaled, 
      MSP_genes_id_corresp %>% 
        select(msp_name,module_name,gene_name_v2) %>% 
        filter(msp_name %in% c("msp_0029")),
      by.x="genes.id", by.y="gene_name_v2") -> msp_counts



binarize_MSP(msp_counts=msp_counts, cutoff_med = 2) -> binary_msp

partitionize_MSP(binary_msp=binary_msp) -> clusters_msp


plot_MSP(binary_msp=binary_msp, 
         clusters_msp=clusters_msp$sample, 
         clusters_msp_gene = clusters_msp$gene)


test_MSP_gene(binary_msp=binary_msp, 
         clusters_msp=clusters_msp$sample) -> test_msp


#clusters_msp %>% head()

plot_MSP(binary_msp=binary_msp %>% 
           merge(test_msp, by="genes.id")  %>% 
           #merge(., kegg_sqlite, by.x="genes.id", by.y="genes_id", all.x=TRUE) %>%
           #merge(., og_sqlite,   by.x="genes.id", by.y="genes_id", all.x=TRUE) %>%
           #filter(!is.na(KO)) %>%
           filter(fdr < 0.05) %>% 
           #select(-p.value, -fdr, -KO, -kegg_genes) , 
           select(-p.value, -fdr),
         clusters_msp=clusters_msp$sample, 
         clusters_msp_gene = clusters_msp$gene) 



  Bstecoris = list(binary_msp=binary_msp, clusters_msp=clusters_msp, test_msp=test_msp  ); save(Bstecoris, file="Bstecoris.rda")
  #Erectale  = list(binary_msp=binary_msp, clusters_msp=clusters_msp, test_msp=test_msp  ); save(Erectale,  file="Erectale.rda")



```

```{r}

binary_msp %>%
  filter(module_name != "core") %>%
  reshape2::melt(id.vars=c("genes.id","module_name")) %>%
  merge(., clusters_msp$sample$partition %>%
          as.matrix %>%
          as.data.frame %>%
          tibble::rownames_to_column("variable") %>%
          dplyr::rename(id=V1),
        by=c("variable")) %>%
  group_by(genes.id) %>%
  mutate(s=sum(value),l=length(value)) %>%
  filter(s>0, s<l) %>%
  group_by(genes.id,id) %>%
  summarize(value=sum(value)) %>%
  reshape2::dcast(genes.id~id) -> genes_prevalence_by_infraspecies


test_msp %>%
  filter(fdr<0.05) %>%
  merge(., kegg_sqlite, by.x="genes.id", by.y="genes_id", all.x=TRUE) %>%
  merge(., og_sqlite,   by.x="genes.id", by.y="genes_id", all.x=TRUE) %>%
  filter(!is.na(OG)) %>%
  merge(genes_prevalence_by_infraspecies, by.x="genes.id") %>%
  arrange(`1`-`2`)






```


## infraspecies of the most prevalent MSP

```{r}
#library(foreach)
#library(doMC)
#doMC::registerDoMC(cores=8)

prevalent_MSP = 
MSP_abundance %>% 
  reshape2::melt(id.vars="msp_name") %>%
  group_by(msp_name) %>%
  filter(value > 0) %>%
  summarize(n=n()) %>%
  mutate(p = n/(dim(MSP_abundance)[2]-1)) %>%
  filter(p>0.25) %>%
  merge(.,MSP_tax, by.y = "MSP_ID", by.x = "msp_name") %>%
  arrange(n)

msp_name_target = prevalent_MSP$msp_name



infraspecies_abundance = NULL

#foreach(i = 1:length(msp_name_target)[1:3] , .combine = rbind) %do% {

pb = txtProgressBar(min = 0, max=length(msp_name_target))

for(i in (1:length(msp_name_target)))  {

setTxtProgressBar(pb,i)

  #i=24  
  
merge(counts_rescaled, 
      MSP_genes_id_corresp %>% 
        select(msp_name,module_name,gene_name_v2) %>% 
        filter(msp_name %in% msp_name_target[i]),
      by.x="genes.id", by.y="gene_name_v2") -> msp_counts

if(dim(msp_counts)[1] < 100 ) next

binarize_MSP(msp_counts=msp_counts, cutoff_med = 2) -> binary_msp

partitionize_MSP(binary_msp=binary_msp) -> clusters_msp

convert_MSP_to_infraspecies(MSP_abundance = MSP_abundance, clusters_msp_sample = clusters_msp$sample, msp_counts = msp_counts) -> tmp

infraspecies_abundance = rbind(infraspecies_abundance,tmp)


}

close(pb)

#} -> infraspecies_abundance

write.csv2(infraspecies_abundance, file="infraspecies_abundance.csv")

```

### explore infraspecies

185 MSP selected and 166 partionized

```{r}

# number of MSP
MSP_abundance %>% dim %>% .[1]

# prevalent MSP
prevalent_MSP$msp_name %>% length

# MSP partitionized
infraspecies_abundance %>%
  filter(partition ==  "cl_2") %>%
  select(msp_name) %>%
  unique %>% .$msp_name -> msp_partitionized

msp_partitionized %>% length()

infraspecies_abundance %>%
  filter(!(msp_name %in%  msp_partitionized)) %>%
  select(msp_name) %>%
  unique %>% .$msp_name -> msp_non_partitionized


infraspecies_abundance %>%
  filter(msp_name %in%  msp_partitionized)  %>%
  mutate(msp_name_partition = paste0(msp_name,"_",partition)) %>%
  reshape2::dcast(msp_name_partition+msp_name~variable, value.var="value", fill = 0) -> infraspecies_table


MSP_abundance %>% 
  filter(!(msp_name %in% msp_partitionized)) %>%
  mutate(msp_name_partition = paste0(msp_name,"_","unassigned")) %>%
  rbind(.,infraspecies_table) %>%
  merge(., MSP_tax, by.x="msp_name", by.y = "MSP_ID" ) -> infraspecies_table_full


infraspecies_table_full

write.csv2(infraspecies_table_full, file="infraspecies_table_full.csv")


```





