---
title: "IBS and food metagenome functional analysis"
output: html_notebook
---

```{r, message=FALSE, warning=FALSE}
library(ggplot2)
library(reshape2)
library(magrittr)
library(ade4)
library(dplyr)
devtools::load_all()
library(IBSMicrobiota)


```


## load data

- co-inertia analysis and supplemental predicted data
- MSP gut modules data by GoMixer
- infraspecies data

```{r, message=FALSE, warning=FALSE}

data(IBSData)
data("nutriscore_DI")

#meat_plant_ratio = read.csv2(file="../inst/tables/meat_plant_ratio.csv", row.names=1)

meat_plant_ratio = readr::read_csv2(system.file("tables", "meat_plant_ratio.csv", package="IBSFood"))[,-1]


load(system.file("data-raw","infraspecies_food_coi_foodsup.rda", package = "IBSFood"))
load(system.file("data-raw","infraspecies_food_coi_spesup.rda",  package = "IBSFood"))
load(system.file("data-raw","infraspecies_food_coi.rda",         package = "IBSFood"))

infraspecies = readr::read_csv2(system.file("infraspecies", "infraspecies_table_full.csv", package="IBSFood"))[,-1]

# MSP_KO_counts=
# read.csv2(system.file("data-raw","MSP_KO_counts.csv",         package = "IBSFood"), row.names = 1)
# 
# KO_counts=
# read.csv2(system.file("data-raw","KO_counts.csv",         package = "IBSFood"), row.names = 1)

# from gomixer
load(system.file("data-raw","modules.rda",         package = "IBSFood"))

load(system.file("data-raw","msp_ko_gut_modules.rda",         package = "IBSFood"))


total_reads = read.csv2(system.file("data-raw","total_reads.csv",         package = "IBSFood"), row.names = 1)

db <- loadDefaultDB() #load modules annotation


```





## check coinertia

It is important to note that Food PCoA Axis 2 associated with degraded food quality
and symptoms severity is negatively correlated with the Axis 2 of the co-inertia. 

Axis 2 of the co-inertia is then correlated with a **BETTER** food quality.


```{r}
infraspecies_food_coi
s.corcircle(infraspecies_food_coi$aY, yax=3)
s.corcircle(infraspecies_food_coi$aX, yax=5)



```



```{r}
# Axis 2 and Axis 4 are linked to gas metabolism, food quality and symptoms severity

microbiota_coinertia = 
rbind(infraspecies_food_coi$lX,infraspecies_food_coi_spesup$lisup) %>%
  select(AxcX1, AxcX2,AxcX4)



```

So Axis 2 and Axis 4 from the co-inertia analysis are respectively associated positively and negatively with a lower H2/CH4 ratio


```{r}

rbind(cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  ggplot() + geom_point(aes(x=H2, y=CH4))

rbind(cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  ggplot() + geom_boxplot(aes(y=AxcX4, x=SSgroup))


rbind(cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  with(.,cor.test(transit.days, AxcX4, method="pearson"))
  
  

rbind(cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  mutate(gas_group = ifelse(CH4>10,"methane\nproducers",ifelse(CH4<10 & H2>10, "high H2\nproducers","low exhaled gas\nproducers"))) %>%
  filter(!is.na(H2)) %>%
  ggplot() + geom_boxplot(aes(x=gas_group,y=AxcX4))


rbind(cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  mutate(gas_group = ifelse(CH4>10,"methane\nproducers",ifelse(CH4<10 & H2>10, "high H2\nproducers","low exhaled gas\nproducers"))) %>%
  filter(!is.na(H2)) %>%
  ggplot() + geom_boxplot(aes(x=gas_group,y=AxcX2))



```

transit time is associated with higher gas production (H2 based or CH4 based). 
to do: test higher transit time with CH4 or with H2 regarding symptoms severity

```{r}

rbind(
  cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  mutate(H2_CH4 = log2((H2+1)/(CH4+1))) %>%
  #filter((H2<10&CH4>10)|(H2>10&CH4<10)) %>% 
  ggplot() + geom_point(aes(x=H2_CH4, y=AxcX2, col=H2>10))

rbind(
  cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  mutate(H2_CH4 = log2((H2+1)/(CH4+1))) %>%
  #filter((H2<10&CH4>10)|(H2>10&CH4<10)) %>% 
  ggplot() + geom_point(aes(x=H2_CH4, y=AxcX4, col=H2>10))



rbind(
  cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  mutate(H2_CH4 = log2((H2+1)/(CH4+1))) %>%
  with(.,cor.test(.$H2_CH4,.$AxcX2, method="spearman"))

rbind(
  cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  mutate(H2_CH4 = log2((H2+1)/(CH4+1))) %>%
  with(.,cor.test(.$H2_CH4,.$AxcX4, method="spearman"))


rbind(
  cbind(group="train",infraspecies_food_coi$lX),
  cbind(group="test",infraspecies_food_coi_spesup$lisup))[1:6] %>% 
  merge(.,
        IBSMicrobiota::IBSData$metadata %>%
           filter(Visit=="V4",Sample_type=="Stool"), 
        by.x="row.names", 
        by.y="Patient_ID") %>%
  mutate(H2_CH4 = log2((H2+1)/(CH4+1))) %>%
  #filter((H2<10&CH4>10)|(H2>10&CH4<10)) %>% 
  ggplot() + geom_point(aes(x=H2_CH4, y=AxcX4-AxcX2, col=H2>10))


```




## gut modules correlations


### correlation to coinertia axis

normalisation : original count where already normalized by gene length and scaled, here we computed proportion
```{r}

modules_global = 
  cbind(
    mods@annotation, 
    sweep(mods@abundance, 2, total_reads[mods2@abundance %>% colnames(),], "/")
    )

modules_per_MSP = 
  cbind(
    mods2@annotation, 
    sweep(mods2@abundance, 2, total_reads[mods2@abundance %>% colnames(),], "/")
    )

modules_per_MSP_coverage = 
cbind(mods2@annotation, mods2@coverage)

m_df =

merge(
  modules_per_MSP %>% melt(id.vars=c("Taxon","Module")),
  modules_per_MSP_coverage %>% melt(id.vars=c("Taxon","Module")), 
  by=c("Taxon","Module","variable")
  ) %>%
  mutate(v=ifelse(value.y> 0.5, value.x,0 )) %>%
  select(Taxon,Module,variable,v) %>%
  reshape2::dcast(Taxon+Module~variable)

modules_per_MSP =
m_df %>%
  mutate_if(is.character, funs(as.numeric))



```


```{r, message=FALSE, warning=FALSE}


modules_global %>%
  melt(id.vars="Module") %>%
  merge(.,microbiota_coinertia, by.x="variable",by.y="row.names") %>%
  group_by(Module) %>%
  do(
    ax1_cor=with(., cor.test(.$value,.$AxcX1, method="spearman") %>% broom::tidy()),  
    ax2_cor=with(., cor.test(.$value,.$AxcX2, method="spearman") %>% broom::tidy()),  
    ax4_cor=with(., cor.test(.$value,.$AxcX4, method="spearman") %>% broom::tidy())
    ) %>%
  tidyr::unnest(ax1_cor, ax2_cor,ax4_cor, .drop = TRUE) %>%
  dplyr::rename(ax1_cor=estimate, ax1_p = `p.value`, ax2_cor=estimate1, ax2_p = `p.value1`, ax4_cor = estimate2, ax4_p = `p.value2`) %>%
  select(Module, ax1_cor, ax1_p, ax2_cor, ax2_p, ax4_cor, ax4_p) %>%
  filter(ax1_p < 0.05 | ax2_p < 0.05 | ax4_p < 0.05) -> modules_cor

module_global_name = 
apply(modules_cor, 1, function(x)getNames(db, x[1]))
#%>%
 # group_by(Module) %>%
  #do(annot=with(.,getNames(db, .$Module ))) %>%
  #tidyr::unnest(annot)

```




```{r, message=FALSE, warning=FALSE}


modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,microbiota_coinertia, by.x="variable",by.y="row.names") %>%
  group_by(Taxon,Module) %>%
  do(
    ax1_cor=with(., cor.test(.$value,.$AxcX1, method="spearman") %>% broom::tidy()),  
    ax2_cor=with(., cor.test(.$value,.$AxcX2, method="spearman") %>% broom::tidy()),  
    ax4_cor=with(., cor.test(.$value,.$AxcX4, method="spearman") %>% broom::tidy())
    ) %>%
  tidyr::unnest(ax1_cor, ax2_cor,ax4_cor, .drop = TRUE) %>%
  dplyr::rename(ax1_cor=estimate, ax1_p = `p.value`, ax2_cor=estimate1, ax2_p = `p.value1`, ax4_cor = estimate2, ax4_p = `p.value2`) %>%
  select(Taxon,Module, ax1_cor, ax1_p, ax2_cor, ax2_p, ax4_cor, ax4_p) %>%
  filter(ax1_p < 0.05 | ax2_p < 0.05 | ax4_p < 0.05) -> MSP_modules_cor

module_name = 
apply(MSP_modules_cor, 1, function(x)getNames(db, x[2]))

  
save(MSP_modules_cor, file="MSP_modules_cor.rda")

```




```{r}

modules_cor %>%
  cbind(module_global_name) %>%
  filter( abs(ax2_cor) > 0.2) %>%
  arrange(desc(ax2_cor)) %>%
  select(module_global_name,ax2_cor)

modules_cor %>%
  cbind(module_global_name) %>%
  filter( abs(ax4_cor) > 0.2) %>%
  arrange(desc(ax4_cor)) %>%
  select(module_global_name,ax4_cor)




```


```{r}

module_name = 
apply(MSP_modules_cor, 1, function(x)getNames(db, x[2]))



MSP_modules_cor %>%
  cbind(module_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  #filter( abs(ax2_cor) > 0.4 | abs(ax4_cor) > 0.4 ) %>%
  #arrange(desc(ax2_cor))
  filter( abs(ax1_cor) > 0.4) %>%
  arrange(desc(ax1_cor)) %>%
   select(species,ax1_cor,module_name, Module)

MSP_modules_cor %>%
  cbind(module_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter( abs(ax2_cor) > 0.4 ) %>%
  arrange(desc(ax2_cor)) %>%
  select(species,ax2_cor,module_name, Module)


MSP_modules_cor %>%
  cbind(module_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter( abs(ax4_cor) > 0.4 ) %>%
  arrange(desc(ax4_cor)) %>%
  select(species,ax4_cor,module_name, Module)



```



```{r}

MSP_modules_cor %>%
  group_by(Taxon , Module) %>%
  do(annot=with(.,getNames(db, .$Module ))) %>%
  tidyr::unnest(annot)

```



### check H2/CH4 ratio

```{r, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,IBSData$metadata %>% 
          filter(Visit=="V4", Sample_type=="Stool"), by.y="Patient_ID", by.x="variable") %>%
  group_by(Taxon,Module) %>%
  do(H2_CH4_cor=with(., cor.test(.$value,log2((.$H2+1)/(.$CH4+1)), method="spearman") %>% broom::tidy())) %>%
  tidyr::unnest(H2_CH4_cor, .drop=TRUE) %>%
  filter(p.value<0.05) -> modules_H2_CH4_MSP_cor


module_H2_CH4_MSP_cor_name = 
apply(modules_H2_CH4_MSP_cor, 1, function(x)getNames(db, x[2]))

modules_H2_CH4_MSP_cor %>%
  cbind(module_H2_CH4_MSP_cor_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter(abs(estimate) > 0.3) %>%
  select(estimate,module_H2_CH4_MSP_cor_name, species) %>%
  arrange(estimate)


modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,IBSData$metadata %>% 
          filter(Visit=="V4", Sample_type=="Stool"), by.y="Patient_ID", by.x="variable") %>%
  group_by(Taxon,Module) %>%
  do(H2_cor=with(., cor.test(.$value,.$H2, method="spearman") %>% broom::tidy())) %>%
  tidyr::unnest(H2_cor, .drop=TRUE) %>%
  filter(p.value<0.05) -> modules_H2_MSP_cor


module_H2_MSP_cor_name = 
apply(modules_H2_MSP_cor, 1, function(x)getNames(db, x[2]))

modules_H2_MSP_cor %>%
  cbind(module_H2_MSP_cor_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter(abs(estimate) > 0.3) %>%
  select(estimate,module_H2_MSP_cor_name, species) %>%
  arrange(estimate)




modules_global %>%
  melt(id.vars=c("Module")) %>%
  merge(.,IBSData$metadata %>% 
          filter(Visit=="V4", Sample_type=="Stool"), by.y="Patient_ID", by.x="variable") %>%
  group_by(Module) %>%
  do(  H2_CH4_cor=with(., cor.test(.$value,log2((.$H2+1)/(.$CH4+1)), method="spearman") %>% broom::tidy())) %>%
  tidyr::unnest(H2_CH4_cor, .drop=TRUE) %>%
  filter(p.value<0.05) -> modules_H2_CH4_cor

module_H2_CH4_cor_name = 
apply(modules_H2_CH4_cor, 1, function(x)getNames(db, x[1]))


modules_H2_CH4_cor %>%
  cbind(module_H2_CH4_cor_name) %>%
  select(estimate,module_H2_CH4_cor_name) %>%
  arrange(estimate)



modules_global %>%
  melt(id.vars=c("Module")) %>%
  merge(.,IBSData$metadata %>% 
          filter(Visit=="V4", Sample_type=="Stool"), by.y="Patient_ID", by.x="variable") %>%
  group_by(Module) %>%
  do(  H2_cor=with(., cor.test(.$value,.$H2, method="spearman") %>% broom::tidy())) %>%
  tidyr::unnest(H2_cor, .drop=TRUE) %>%
  filter(p.value<0.05) -> modules_H2_cor

module_H2_cor_name = 
apply(modules_H2_cor, 1, function(x)getNames(db, x[1]))


modules_H2_cor %>%
  cbind(module_H2_cor_name) %>%
  select(estimate,module_H2_cor_name) %>%
  arrange(estimate)




```


## check infraspecies modules



```{r, fig.height=5, fig.width=10}


msp_target = "msp_0071"

infraspecies_select=
infraspecies %>% 
  filter(msp_name==msp_target) %>%
  select(contains("M", ignore.case = FALSE), msp_name_partition,msp_name,species) %>%
  melt(id.vars=c("msp_name_partition","msp_name","species")) %>%
  group_by(variable) %>%
  top_n(1)



modules_per_MSP %>%
  filter(Taxon==msp_target) %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(infraspecies_select, by="variable") %>%
  ungroup %>%
  ggplot() + geom_point(aes(x=sqrt(value.x),y=sqrt(value.y),col=msp_name_partition))
  

modules_per_MSP %>%
  filter(Taxon==msp_target) %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(infraspecies_select, by="variable") %>%
  ungroup %>% 
  group_by(msp_name_partition) %>%
  mutate(value= log10(value.x+10^-6) %>% scale) %>%
  filter(!(msp_name_partition %in% grep("unassigned",msp_name_partition, value=TRUE )) ) %>%
  #filter(Module %in% c("MF0001","MF0038")) %>%
  ggplot() + 
  geom_tile(aes(y=Module,x=variable,fill=value)) + 
  facet_wrap(~msp_name_partition, scales = "free_x") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0, size=2), axis.text.y  = element_text(size=4))



modules_per_MSP %>%
  filter(Taxon==msp_target) %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(infraspecies_select, by="variable") %>%
  ungroup %>% 
  group_by(msp_name_partition) %>%
  mutate(value= log10(value.x+10^-6) %>% scale) %>%
  filter(!(msp_name_partition %in% grep("unassigned",msp_name_partition, value=TRUE )) ) %>%
  ggplot() + geom_boxplot(aes(y=value ,x=Module, fill=msp_name_partition)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0, size=10))




```




```{r}

modules_per_MSP = cbind(mods2@annotation, sweep(mods2@abundance, 2, total_reads[mods2@abundance %>% colnames(),], "/"))



modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  group_by(variable) %>%
  summarize(s = sum(value)) %>%
  ggplot() + geom_histogram(aes(x=s %>% as.character %>% as.numeric()))


mods2@abundance %>%
  melt() %>%
  group_by(variable) %>%
  summarize(s = sum(value) %>% round(1)) %>%
  merge(total_reads, by.x="variable", by.y="row.names") %>%
  mutate(p=s/V1) %>%
  arrange(desc(p)) %>%
  
  
  ggplot() + geom_histogram(aes(x=s %>% as.character %>% as.numeric()))


```


## Modules associations on infraspecies PCoA

> TO DO: check axis direction!!



Coinertia Axis 1 is **positively** associated with plant/meat based ratio (food axis 1) and **negatively** with microbiota axis 3.
Coinertia Axis 2 (associated with higher CH4) is **negatively** associated with nutriscore & symptoms (food axis 2) and **positevely** with microbiota axis 2. 
Coinertia Axis 4 (associated with higher H2) is **negatively** associated microbiota axis 1.  microbiota Axis 1 is associated with higher Prevotella and lower bacteroides.


### microbiota axis 1

Microbiota axis 1 is associated with lower H2

```{r, message=FALSE, warning=FALSE}
modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,infraspecies_all_pco$li, by.x="variable",by.y="row.names") %>%
  group_by(Taxon,Module) %>%
  do(
    #ax1_cor=with(., cor.test(.$value,.$A1, method="spearman") %>% broom::tidy()),  
    #ax2_cor=with(., cor.test(.$value,.$AxcX2, method="spearman") %>% broom::tidy()),  
    ax1_cor=with(., cor.test(.$value,.$A1, method="spearman") %>% broom::tidy())
    ) %>%
  tidyr::unnest(ax1_cor,  .drop = TRUE) %>%
  dplyr::rename(ax1_cor=estimate, ax1_p = `p.value`) %>%
  select(Taxon,Module, ax1_cor, ax1_p) %>%
  filter(ax1_p < 0.05) -> MSP_modules_PCO_cor

MSP_modules_PCO_cor_name = 
apply(MSP_modules_PCO_cor, 1, function(x)getNames(db, x[2]))


MSP_modules_PCO_cor %>%
  cbind(name=MSP_modules_PCO_cor_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter(abs(ax1_cor)>0.4) %>%
  arrange(ax1_cor) -> microbiota_associations_with_low_H2

write.csv2(microbiota_associations_with_low_H2,file="microbiota_associations_with_low_H2.csv")



```

### microbiota axis 3

Microbiota axis 3 is associated with higher meat conso.

```{r, message=FALSE, warning=FALSE}

modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,infraspecies_all_pco$li, by.x="variable",by.y="row.names") %>%
  group_by(Taxon,Module) %>%
  do(
    #ax1_cor=with(., cor.test(.$value,.$A1, method="spearman") %>% broom::tidy()),  
    #ax2_cor=with(., cor.test(.$value,.$AxcX2, method="spearman") %>% broom::tidy()),  
    ax3_cor=with(., cor.test(.$value,.$A3, method="spearman") %>% broom::tidy())
    ) %>%
  tidyr::unnest(ax3_cor,  .drop = TRUE) %>%
  dplyr::rename(ax3_cor=estimate, ax3_p = `p.value`) %>%
  select(Taxon,Module, ax3_cor, ax3_p) %>%
  filter(ax3_p < 0.05) -> MSP_modules_PCO_cor

MSP_modules_PCO_cor_name = 
apply(MSP_modules_PCO_cor, 1, function(x)getNames(db, x[2]))


MSP_modules_PCO_cor %>%
  cbind(name=MSP_modules_PCO_cor_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter(abs(ax3_cor)>0.4) %>%
  arrange(ax3_cor)



```

### microbiota axis 2

microbiota axis 2 is associated with higher CH4 and lower nutriscore (so better food quality)

```{r}

modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,infraspecies_all_pco$li, by.x="variable",by.y="row.names") %>%
  group_by(Taxon,Module) %>%
  do(
    #ax1_cor=with(., cor.test(.$value,.$A1, method="spearman") %>% broom::tidy()),  
    #ax2_cor=with(., cor.test(.$value,.$AxcX2, method="spearman") %>% broom::tidy()),  
    ax2_cor=with(., cor.test(.$value,.$A2, method="spearman") %>% broom::tidy())
    ) %>%
  tidyr::unnest(ax2_cor,  .drop = TRUE) %>%
  dplyr::rename(ax2_cor=estimate, ax2_p = `p.value`) %>%
  select(Taxon,Module, ax2_cor, ax2_p) %>%
  filter(ax2_p < 0.05) -> MSP_modules_PCO_cor

MSP_modules_PCO_cor_name = 
apply(MSP_modules_PCO_cor, 1, function(x)getNames(db, x[2]))


MSP_modules_PCO_cor %>%
  cbind(name=MSP_modules_PCO_cor_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter(abs(ax2_cor)>0.4) %>%
  arrange(ax2_cor) -> microbiota_associations_with_higher_CH4_better_food_quality

write.csv2(microbiota_associations_with_higher_CH4_better_food_quality,file="microbiota_associations_with_higher_CH4_better_food_quality.csv")


```


### microbiota axis 4

```{r}

modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,infraspecies_all_pco$li, by.x="variable",by.y="row.names") %>%
  group_by(Taxon,Module) %>%
  do(
    #ax1_cor=with(., cor.test(.$value,.$A1, method="spearman") %>% broom::tidy()),  
    #ax2_cor=with(., cor.test(.$value,.$AxcX2, method="spearman") %>% broom::tidy()),  
    ax4_cor=with(., cor.test(.$value,.$A4, method="spearman") %>% broom::tidy())
    ) %>%
  tidyr::unnest(ax4_cor,  .drop = TRUE) %>%
  dplyr::rename(ax4_cor=estimate, ax4_p = `p.value`) %>%
  select(Taxon,Module, ax4_cor, ax4_p) %>%
  filter(ax4_p < 0.05) -> MSP_modules_PCO_cor

MSP_modules_PCO_cor_name = 
apply(MSP_modules_PCO_cor, 1, function(x)getNames(db, x[2]))


MSP_modules_PCO_cor %>%
  cbind(name=MSP_modules_PCO_cor_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter(abs(ax4_cor)>0.4) %>%
  arrange(ax4_cor)


```




## Modules associations

we will do spearman correlation with:

- H2/CH4
- Nutriscore diet index
- IBS symptoms severity
- Meat/plan ratio


### prepare metadata


```{r, message=FALSE, warning=FALSE}

IBSData$metadata %>%
  filter(Visit=="V4", Sample_type=="Stool") %>%
  select(Patient_ID,Health,H2,CH4,IBS.Severity.Score) %>%
  mutate(`IBS.Severity.Score` = ifelse(Health=="control", 0, `IBS.Severity.Score` )) %>%
  mutate(H2_CH4_ratio = log2((H2+1)/(CH4+1))) %>%
  merge(.,nutriscore_DI, by="Patient_ID", all.x=TRUE) %>%
  merge(.,meat_plant_ratio %>% select(variable,meat_plant_ratio), by.x="Patient_ID", by.y="variable", all.x=TRUE) %>%
  select(-H2,-CH4) %>%
  melt(id.vars=c("Patient_ID","Health")) -> ibs_metadata_select
  


```


### Global Modules


```{r}

modules_global %>% 
  melt(id.vars=c("Module")) %>%
  merge(ibs_metadata_select,., by.y="variable", by.x="Patient_ID") %>%
  group_by(variable,Module) %>%
  do(  rho_cor=with(., cor.test(.$value.x, .$value.y, method="spearman") %>% broom::tidy())) %>%
  tidyr::unnest(rho_cor, .drop=TRUE) %>%
  select(variable,Module,estimate,p.value) -> modules_global_rho_test
  

modules_global_rho_test_name = 
apply(modules_global_rho_test, 1, function(x)getNames(db, x[2]))


modules_global_rho_test$Modules_names = modules_global_rho_test_name
 
```

```{r, fig.height=7, fig.width=9}

modules_global_rho_test %>%
  rename(rho = estimate) %>%
  group_by(Module,Modules_names) %>%
  filter(any(p.value<0.05)) %>%
  ggplot() + geom_tile(aes(y=paste0(Module," - ",Modules_names),x=variable,fill=rho,alpha=-log10(p.value))) + 
  scale_fill_gradient2() + theme_classic() + ylab("") + xlab("")


```

### Modules per MSP

```{r, message=FALSE, warning=FALSE}

data(MSP_tax)


IBSData$metadata %>%
  filter(Visit=="V4", Sample_type=="Stool") %>%
  select(Patient_ID,Health,H2,CH4,IBS.Severity.Score) %>%
  mutate(`IBS.Severity.Score` = ifelse(Health=="control", 0, `IBS.Severity.Score` )) %>%
  mutate(H2_CH4_ratio = log2((H2+1)/(CH4+1))) %>%
  merge(.,nutriscore_DI, by="Patient_ID", all.x=TRUE) %>%
  merge(.,meat_plant_ratio %>% select(variable,meat_plant_ratio), by.x="Patient_ID", by.y="variable", all.x=TRUE) %>%
  select(-H2,-CH4, -Health) %>%
  tibble::column_to_rownames("Patient_ID") -> ibs_metadata_select_transp
  
modules_per_MSP %>%
  mutate(Taxon_modules = paste0(Taxon,"__",Module)) %>% 
  select(-Taxon, -Module) %>%
  tibble::column_to_rownames("Taxon_modules") %>%
  t -> modules_per_MSP_transp
  
modules_per_MSP_metadata_cor = 
cor(modules_per_MSP_transp, ibs_metadata_select_transp[row.names(modules_per_MSP_transp),], method="spearman", use = "complete.obs")


modules_per_MSP_metadata_cor %>%
  melt() %>%
  filter(!is.na(value)) %>% 
  #filter(abs(value)>0.3) %>%
  tidyr::separate(col="Var1", into=c("Taxon","Module"), sep="__"  ) -> modules_per_msp_rho_test


 
modules_per_msp_rho_test_name = 
apply(modules_per_msp_rho_test, 1, function(x)getNames(db, x[2]))


modules_per_msp_rho_test$Modules_names = modules_per_msp_rho_test_name
  
  
modules_per_msp_rho_test %>%
  merge(.,MSP_tax %>% select(MSP_ID, species,phylum), by.x="Taxon", by.y="MSP_ID") -> modules_per_msp_rho_test



```



```{r, fig.height=15, fig.width=30}


modules_per_msp_rho_test %>%
  rename(rho = value, variable = Var2) %>%
  group_by(Module,Taxon,Modules_names,species,phylum) %>%
  filter(any(abs(rho)>0.4)) %>%
  ggplot() + geom_tile(aes(x=paste0(Module," - ",Modules_names),y=paste(species,Taxon),fill=rho)) + 
  facet_wrap(~variable, scales = "free_x", nrow = 1) +
  scale_fill_gradient2() + theme_dark() + ylab("") + xlab("") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.25))

ggsave("gut_modules_per_msp.pdf", h=10, w=20)







```


```{r}

modules_per_msp_rho_test %>%
  rename(rho = value, variable = Var2) %>%
  group_by(Module,Taxon,Modules_names,species,phylum) %>%
  filter(Taxon=="msp_0070") %>%
  arrange(desc(abs(rho)))


```


```{r, fig.height=15, fig.width=30}

modules_per_msp_rho_test %>%
  rename(rho = value, variable = Var2) %>%
  filter(variable %in% c("IBS.Severity.Score", "nutriscore_DI","meat_plant_ratio")) %>%
  group_by(Module,Taxon,Modules_names,species,phylum) %>%
  filter(any(abs(rho)>0.25)) %>%
  ungroup() %>%
  group_by(Module,Modules_names, variable) %>%
  mutate(n=n()) %>% 
  filter(n>=2) %>%
  filter(all(abs(rho)> 0.25))  %>%
  ggplot() + geom_tile(aes(x=paste0(Module," - ",Modules_names),y=paste(species),fill=rho)) + 
  facet_grid(phylum~variable, scales = "free_y", drop = TRUE, space="free") +
  scale_fill_gradient2() + theme_dark() + ylab("") + xlab("") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.25))




```




```{r}


modules_per_msp_rho_test %>%
  rename(rho = value, variable = Var2) %>%
  group_by(Module,Taxon,Modules_names) %>%
  filter(any(abs(rho)>0.4)) %>%
  ggplot() + geom_tile(aes(x=variable,y=Taxon,fill=rho)) + 
  facet_wrap(~paste0(Module," - ",Modules_names), scales = "free_y") +
  scale_fill_gradient2() + theme_dark() + ylab("") + xlab("") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```




#### circos plot

```{r}

par(mar=c(5,5,5,5))

circos.clear()


circos.par(cell.padding=c(0,0,0,0), track.margin=c(0,0), start.degree = 0, gap.degree = 1, points.overflow.warning=FALSE)

chordDiagram(na.omit(data.frame(test$ClusterDTU.x, test$egg)[order(test[,3]),]), col= c("red","green")[as.numeric(test[order(test[,3]),3]>0)+1], annotationTrack= "grid")

circos.trackPlotRegion(track.index = 1, panel.fun = function(x, y) {
    xlim  = get.cell.meta.data("xlim")
    xplot = get.cell.meta.data("xplot")
    ylim  = get.cell.meta.data("ylim")
    sector.name = get.cell.meta.data("sector.index")

    if(abs(xplot[2] - xplot[1]) < 80) {
        circos.text(mean(xlim), ylim[1], sector.name, facing = "clockwise",
            niceFacing = TRUE, adj = c(0, 0.5), cex=0.5)
    } else {
        circos.text(mean(xlim), ylim[1], sector.name, facing = "inside", 
            niceFacing = TRUE, adj = c(0.5, 0), cex=0.5)
    }
}, bg.border = NA)


```




## Misc


### Modules Global



```{r}


modules_global %>%
  melt(id.vars=c("Module")) %>%
  merge(.,IBSData$metadata %>% 
          filter(Visit=="V4", Sample_type=="Stool"), by.y="Patient_ID", by.x="variable") %>%
  group_by(Module) %>%
  do(  H2_CH4_cor=with(., cor.test(.$value,log2((.$H2+1)/(.$CH4+1)), method="spearman") %>% broom::tidy())) %>%
  tidyr::unnest(H2_CH4_cor, .drop=TRUE) %>%
  filter(p.value<0.05) -> modules_H2_CH4_cor

module_H2_CH4_cor_name = 
apply(modules_H2_CH4_cor, 1, function(x)getNames(db, x[1]))


modules_H2_CH4_cor %>%
  cbind(module_H2_CH4_cor_name) %>%
  select(estimate,module_H2_CH4_cor_name) %>%
  arrange(estimate)




```


### Modules per MSP

```{r}

modules_per_MSP %>%
  melt(id.vars=c("Taxon","Module")) %>%
  merge(.,IBSData$metadata %>% 
          filter(Visit=="V4", Sample_type=="Stool"), by.y="Patient_ID", by.x="variable") %>%
  group_by(Taxon,Module) %>%
  do(H2_CH4_cor=with(., cor.test(.$value,log2((.$H2+1)/(.$CH4+1)), method="spearman") %>% broom::tidy())) %>%
  tidyr::unnest(H2_CH4_cor, .drop=TRUE) %>%
  filter(p.value<0.05) -> modules_H2_CH4_MSP_cor


module_H2_CH4_MSP_cor_name = 
apply(modules_H2_CH4_MSP_cor, 1, function(x)getNames(db, x[2]))

modules_H2_CH4_MSP_cor %>%
  cbind(module_H2_CH4_MSP_cor_name) %>%
  merge(infraspecies %>% select(msp_name,species, genus) %>% unique, by.x="Taxon", by.y="msp_name") %>%
  filter(abs(estimate) > 0.3) %>%
  select(estimate,module_H2_CH4_MSP_cor_name, species) %>%
  arrange(estimate)




```




## Modules MSP IBS severity



```{r}

MSP_severity = read.csv2("MSP_severity.csv", row.names=1)




modules_per_MSP %>%
  select(Taxon,Module) %>%
  merge(.,MSP_severity %>% select(Var2,q50,species.y), by.x="Taxon", by.y="Var2") -> modules_per_MSP_severity


modules_per_MSP_severity_name = 
apply(modules_per_MSP_severity, 1, function(x)getNames(db, x[2]))


modules_per_MSP_severity$Module_name = modules_per_MSP_severity_name




```





```{r, fig.height=15, fig.width=15}


modules_per_MSP_severity %>%
  mutate(IBSseverityMSP = ifelse(q50>0,"Healthy","Severe")) %>%
  ggplot() + geom_tile(aes(y=paste(Module,Module_name), x=species.y)) + facet_wrap(~ IBSseverityMSP, scales = "free_x") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
  
  


```


## global picture


```{r, message=FALSE, warning=FALSE}

data("MSP_tax")

MSP_abundance = read.csv2( file="MSP_abundance.csv", row.names=2)[,-1]

MSP_dominant = 
MSP_abundance %>% as.matrix() %>% 
  prop.table(2) %>% as.data.frame %>% 
  BiotypeR::noise.removal(percent=1) %>%
  rownames()

MSP_severity = read.csv2(file="MSP_severity.csv", row.names=1)


MSP_tax %>%
  filter(MSP_ID %in% MSP_dominant)


egg = readr::read_tsv(system.file("data-raw","egg_db_updated.tsv.txt",         package = "IBSFood")) #updated

modules_structures = readr::read_tsv(system.file("data-raw","gut_modules_names_structure.tsv",         package = "IBSFood")) 



load(system.file("data-raw","msp_ko_gut_modules.rda",         package = "IBSFood"))

modules_per_MSP = 
  cbind(
    mods2@annotation, 
    sweep(mods2@abundance, 2, total_reads[mods2@abundance %>% colnames(),], "/")
    )



modules_per_MSP_name = 
apply(modules_per_MSP, 1, function(x)getNames(db, x[2]))


modules_per_MSP$Module_name = modules_per_MSP_name



# modules_per_MSP =
# 
# cbind(
#   Module      = modules_per_MSP %>% select(2) %>% unique(),
#   Module_name = modules_per_MSP %>% select(2) %>% unique() %>% apply(., 1, function(x)getNames(db, x[1]))
#   ) %>%
#   merge(modules_per_MSP,., by="Module")
# 


```




```{r}


modules_per_MSP_annotation =

  modules_per_MSP %>% 
  melt(id.vars=c("Module","Taxon","Module_name")) %>%
  filter(value > 10^-9 ) %>%
  group_by(Module,Taxon,Module_name) %>%
  summarise(n = n()) %>%
  ungroup() %>%
  group_by(Taxon) %>%
  mutate(p= n/max(n)) %>% # prevalence of modules within msp
  filter(Taxon %in% c(MSP_dominant,MSP_severity$Var2 %>% as.character)) %>%
  merge(MSP_tax,.,by.x="MSP_ID", by.y="Taxon") %>%
  merge(., MSP_severity %>% select(Var2,q50), by.x="MSP_ID", by.y="Var2", all.x=TRUE ) %>%
  merge(.,egg, by.x="Module", by.y="module_id") %>%
  mutate(severe = ifelse(is.na(q50),0,q50))

  
egg



```



```{r}

modules_per_MSP_annotation %>%
  filter(MSP_ID=="msp_0492")

MSP_tax %>%
  filter(MSP_ID=="msp_a10")


modules_per_MSP %>%
  filter(Taxon=="msp_0492")



```


```{r, fig.height=15, fig.width=20}

modules_per_MSP_annotation %>%
  filter(q50!=0) %>%
  #filter(`Hierarchy Level 1` %in% c("gas metabolism","organic acid metabolism","alcohol metabolism")) %>%
  #filter(`Hierarchy Level 1` %in% c("carbohydrate degradation")) %>%
  #filter(`Hierarchy Level 1` %in% c("amino acid degradation","glycoprotein degradation","amines and polyamines degradation","lipid degradation", "inorganic nutrient metabolism")) %>%
  filter(phylum %in% c("Bacteroidetes","Firmicutes","Actinobacteria","Proteobacteria","verrucomicrobia","Euryarchaeota")) %>%
  ggplot() + 
  geom_tile(aes(x=module_name, y=paste0(species," ",MSP_ID), alpha=p, fill=severe)) + 
  scale_fill_gradient2(low = "red", mid="yellow", high="green") +
  facet_grid(q50>0~`Hierarchy Level 2` %>% gsub(" ", "\n", .),scales = "free", space = "free") + 
   theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))




```


```{r}


matrix(c(12,22,6,22),nr=2) %>% t %>% chisq.test()



```


```{r, message=FALSE, warning=FALSE}

MSP_severity

merge(nutriscore_DI,meat_plant_ratio,by.x="Patient_ID", by.y="variable") %>%
  select(Patient_ID, nutriscore_DI,meat_plant_ratio) %>%
  merge(., 
        IBSData$metadata %>% filter(Visit == "V4", Sample_type == "Stool") %>% select(Patient_ID, H2,CH4),
        by="Patient_ID") %>%
  merge(., MSP_abundance[MSP_severity$Var2 %>% as.character,] %>% 
  as.matrix() %>%
  reshape::melt() , by.x="Patient_ID", by.y="X2" ) %>%
  group_by(X1) %>%
  do(
    cor.test(.$meat_plant_ratio, .$value, method="spearman", data=.) %>% broom::tidy()
  ) %>% 
  merge(MSP_tax,., by.x="MSP_ID", by.y="X1") %>%
arrange(p.value)





```



```{r}

diet_severity_df =

merge(nutriscore_DI,meat_plant_ratio,by.x="Patient_ID", by.y="variable") %>%
  select(Patient_ID, nutriscore_DI,meat_plant_ratio) %>%
  merge(., 
        IBSData$metadata %>% 
          filter(Visit == "V4", Sample_type == "Stool") %>% 
          select(Patient_ID, H2,CH4) ,
        by="Patient_ID") %>% 
  merge(., MSP_abundance[MSP_severity$Var2 %>% as.character,] %>% 
  as.matrix() %>%
  reshape::melt() , by.x="Patient_ID", by.y="X2" ) %>%
  reshape2::dcast(Patient_ID+nutriscore_DI+meat_plant_ratio+H2+CH4~X1) %>%
  tibble::column_to_rownames("Patient_ID")



dudi.pca(diet_severity_df %>% 
           na.omit() %>%
    select(contains("msp")) %>% 
    add(10^-9) %>% log10() , scannf=FALSE, nf=3) -> severity_pca
  
  pcaiv(severity_pca, df = diet_severity_df %>% na.omit() %>%
    select(nutriscore_DI, meat_plant_ratio, H2, CH4), nf=3, scannf=FALSE
      ) -> diet_severity_pcaiv


  diet_severity_pcaiv$c1 %>%
    merge(MSP_tax,., by.y="row.names", by.x="MSP_ID") %>%
    ggplot() + geom_text(aes(x=CS1*2,y=CS2*2, label=species), size=2) + geom_text(data=diet_severity_pcaiv$cor, aes(x=RS1/2,y=RS2/2, label=row.names(diet_severity_pcaiv$cor)), col="red")
    
  

```

### heatmap
```{r, fig.height=10, fig.width=15, message=FALSE, warning=FALSE}


load("food_otu.rda")

food_MSP_severity_cor =
food_otu %>%
  prop.table(2) %>%
  melt() %>%
  dplyr::rename(Subject=Var2, food_group=Var1) %>%
  merge(.,
    MSP_abundance[MSP_severity$Var2 %>% as.character,] %>%
      as.matrix() %>%
      melt() %>%
      dplyr::rename(Subject=Var2, MSP_id=Var1) %>%
      group_by(MSP_id,Subject,value) %>% unique(), by="Subject"
    ) %>%
  group_by(food_group,MSP_id) %>%
  summarise(rho=cor(value.x,value.y,method="spearman")) %>%
  na.omit() 
  



```

```{r, fig.height=10, fig.width=18}



library(pheatmap)
  
food_MSP_severity_heatmap =  
  food_MSP_severity_cor %>%
  merge(.,MSP_severity %>% select(species.y,Var2) %>% unique, by.x="MSP_id", by.y="Var2") %>%
  select(food_group,species.y,MSP_id,rho) %>%
  mutate(MSP_id = paste(MSP_id,species.y)) %>%
  select(food_group, MSP_id,rho) %>%
  reshape2::dcast(food_group~MSP_id) %>% 
  tibble::column_to_rownames("food_group") %>%
  as.matrix() %>% t 


data(nutriscore_food_group)

nutriscore_food_group =
  nutriscore_food_group %>%
  #tibble::column_to_rownames("Food groups lvl4") %>%
  select(nutriscore, `Food groups lvl1`, `Food groups lvl2`, `Food groups lvl4`) %>%
  mutate(meat_plant = ifelse(`Food groups lvl2` %in% c("Meat","Fish"), "Meat-based",  "Others" )) %>%
  mutate(meat_plant = ifelse(`Food groups lvl1` %in% "Plant-based", "Plant-based", meat_plant  )) %>%
  tibble::column_to_rownames("Food groups lvl4") %>%
  select(nutriscore,meat_plant)


MSP_severity_q50 =
MSP_severity %>%
  mutate(MSP_id = paste(Var2,species.y)) %>%
  filter(MSP_id %in% rownames(food_MSP_severity_heatmap)) %>%
  select(MSP_id,q50) %>%
  group_by(MSP_id) %>%
  summarize(q50 = mean(q50)) %>%
  mutate(q50=ifelse(q50>0,"Healthy","Severe")) %>%
  as.data.frame() %>%
  tibble::column_to_rownames("MSP_id")

ann_colors = list(
    meat_plant = c(`Meat-based`="pink", Others="grey", `Plant-based`="green"),
    q50 = c(Healthy="light blue", Severe="brown")#,
    #GeneClass = c(Path1 = "#7570B3", Path2 = "#E7298A", Path3 = "#66A61E")
)

pheatmap_plot=

pheatmap(food_MSP_severity_heatmap, clustering_distance_rows = "euclidean", clustering_distance_cols = "euclidean", cutree_rows = 4, cutree_cols = 4, annotation_col = nutriscore_food_group, annotation_row = MSP_severity_q50, annotation_colors = ann_colors, clustering_method = "ward.D2" ) 
  
print(pheatmap_plot)

  pheatmap(food_MSP_severity_heatmap, clustering_distance_rows = "euclidean", clustering_distance_cols = "euclidean", cutree_rows = 4, cutree_cols = 4, annotation_col = nutriscore_food_group, annotation_row = MSP_severity_q50 ) %>% unclass

  
  
  # merge(.,MSP_severity, by.x="MSP_id", by.y="Var2") %>%
  # ggplot() + geom_tile(aes(x=food_group,y=MSP_id, fill=rho)) + 
  # facet_wrap(~q50>0, scale = "free", nrow=2) +
  # scale_fill_gradient2(mid="yellow", high="green") +
  # theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5)) 
```





