---
title: "large correlation analysis"
output: html_notebook
---



```{r}

library(dplyr)
library(magrittr)
library(ggplot2)
devtools::load_all()

```


```{r, message=FALSE, warning=FALSE}


infraspecies = readr::read_csv2(system.file("infraspecies", "infraspecies_table_full.csv", package="IBSFood"))[,-1]

nutrient_taxa_cor = read.csv2(file="nutrient_taxa_cor.csv", row.names=1)

meat_plant_ratio = readr::read_csv2(system.file("tables", "meat_plant_ratio.csv", package="IBSFood"))[,-1]


```


## prepare compo table

filter low abundant subspecies


```{r}

infraspecies %>% dim


infraspecies %>%
  filter(genus!="Homo") %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum,class,order,family,genus,msp_name) %>%
  group_by(superkingdom,phylum,class,order,family,genus,msp_name) %>%
  summarise_all(sum) %>%
  as.data.frame() %>%
  tibble::column_to_rownames("msp_name") %>%
  select(contains("M", ignore.case = F)) %>%
  BiotypeR::noise.removal(percent=0.01) %>% row.names -> species_select

species_select %>% length()

# infraspecies %>%
#   filter(genus!="Homo") %>%
#   mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
#   as.data.frame() %>%
#   tibble::column_to_rownames("msp_name_partition") %>%
#   select(contains("M", ignore.case = F)) %>%
#   BiotypeR::noise.removal(percent=0.1) %>% row.names -> infraspecies_select
  #summarise_all(funs(sum)) %>% t %>%
  #summary


# infraspecies_select %>% head
# 
# infraspecies_select = infraspecies_select[grep("unassigned", infraspecies_select, invert = TRUE)]

infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  summarise_at(vars(contains("M", ignore.case = F)),function(x){sum(x)}) %>%
  t %>% as.data.frame %>% ggplot() + geom_density(aes(x=V1)) + xlim(0.8,1) 


```




melt composition table at each tax level
```{r}



rbind(
  
  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom) %>%
  group_by(superkingdom) %>%
  summarise_all(sum) %>%
    ungroup() %>%
  reshape2::melt(id.vars="superkingdom") %>%
  mutate(taxa = "superkingdom") %>%
  dplyr::rename(tax_variable = superkingdom),
  
  
  
  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum) %>%
  group_by(superkingdom,phylum) %>%
  summarise_all(sum) %>%
    ungroup() %>%
  select(-superkingdom) %>%
  reshape2::melt(id.vars="phylum") %>%
  mutate(taxa = "phylum") %>%
  dplyr::rename(tax_variable = phylum),

  
  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum,class) %>%
  group_by(superkingdom,phylum,class) %>%
  summarise_all(sum) %>%
    ungroup() %>%
  select(-superkingdom,-phylum) %>%
  reshape2::melt(id.vars="class") %>%
  mutate(taxa = "class") %>%
  dplyr::rename(tax_variable = class),
  
  
  
  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum,class,order) %>%
  group_by(superkingdom,phylum,class,order) %>%
  summarise_all(sum) %>%
    ungroup() %>%
  select(-superkingdom,-phylum,-class) %>%
  reshape2::melt(id.vars="order") %>%
  mutate(taxa = "order") %>%
  dplyr::rename(tax_variable = order),
  
  
  
  
  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum,class,order,family) %>%
  group_by(superkingdom,phylum,class,order,family) %>%
  summarise_all(sum) %>%
  ungroup() %>%
  select(-superkingdom,-phylum,-class,-order) %>%
  reshape2::melt(id.vars="family")%>%
  mutate(taxa = "family") %>%
  dplyr::rename(tax_variable = family),


  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum,class,order,family,genus) %>%
  group_by(superkingdom,phylum,class,order,family,genus) %>%
  summarise_all(sum) %>%
  ungroup() %>%
  select(-superkingdom,-phylum,-class,-order,-family) %>%
  reshape2::melt(id.vars="genus") %>%
  mutate(taxa = "genus") %>%
  dplyr::rename(tax_variable = genus),


  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum,class,order,family,genus,msp_name) %>%
  group_by(superkingdom,phylum,class,order,family,genus,msp_name) %>%
  summarise_all(sum) %>%
  ungroup() %>%
  select(-superkingdom,-phylum,-class, -order, -family,-genus) %>%
  reshape2::melt(id.vars="msp_name") %>%
  mutate(taxa = "MSP") %>%
  dplyr::rename(tax_variable = msp_name),


  infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(msp_name %in% species_select) %>%
  select(contains("M", ignore.case = F),superkingdom,phylum,class,order,family,genus,msp_name, msp_name_partition) %>%
  group_by(superkingdom,phylum,class,order,family,genus,msp_name, msp_name_partition) %>%
  summarise_all(sum) %>%
  ungroup() %>%
  select(-superkingdom,-phylum,-class, -order, -family,-genus, -msp_name) %>%
  reshape2::melt(id.vars="msp_name_partition") %>%
  mutate(taxa = "Infraspecies") %>%
  dplyr::rename(tax_variable = msp_name_partition) #%>%
  #mutate(unassigned =  grepl("unassigned", tax_variable)   )  %>%
  #  filter(!unassigned) %>% select(-unassigned)
    
    
) -> compo_taxa_melt

save(compo_taxa_melt, file="compo_taxa_melt.rda")


compo_taxa_melt %>% head



```



## correlation with nutrient

```{r eval=FALSE, include=FALSE}


data("nutrients_data")

nutrients_data %>%
  reshape2::melt(id.vars="ID") %>%
  merge(.,compo_taxa_melt, by.y="variable", by.x="ID") -> compo_taxa_nutrient_melt
  


compo_taxa_nutrient_melt %>% 
  #filter(variable == "Niek _mg" | tax_variable == "msp_1265_unassigned" | tax_variable =="msp_0849_cl_1"| tax_variable == "msp_0849") %>%
  group_by(taxa,tax_variable,variable) %>%
  summarize(cor = cor(value.x,value.y, method="spearman")) -> nutrient_taxa_cor




```



## large permanova with food unifrac

```{r, message=FALSE, warning=FALSE}
load("food_otu.rda")

load("food_unifrac.dist")




id = intersect(attr(food_unifrac, "Label"), colnames(infraspecies %>% select(contains("M", ignore.case = F))))


compo_taxa_food_unifrac_permanova =

compo_taxa_melt %>%
  filter(variable %in% id) %>%
  #filter(taxa == "family") %>%
  mutate(variable = variable %>% as.character) %>%
  arrange(variable) %>%
  group_by(tax_variable,taxa) %>%
  do(vegan::adonis2(food_unifrac %>% as.matrix %>% .[sort(id),sort(id)] %>% as.dist  ~ value , data=. ) %>% broom::tidy() %>% slice(1)  ) %>%
  arrange(p.value)

write.csv2(compo_taxa_food_unifrac_permanova,file = "compo_taxa_food_unifrac_permanova.csv")  
#compo_taxa_food_unifrac_permanova %>% ggplot() + geom_point(aes(x=R2,y=-log10(p.value), color=p.value<0.05))
  
compo_taxa_food_unifrac_permanova %>%
  filter(tax_variable=="msp_0070_cl_1")

infraspecies %>%
  filter(msp_name_partition == "msp_0070_cl_1")


infraspecies %>%
  filter(species == "Eubacterium rectale")

```

## large correlation with meat_plant ratio


```{r, message=FALSE, warning=FALSE}


meat_plant_ratio %>% select(variable,meat_plant_ratio)

id = intersect(attr(food_unifrac, "Label"), colnames(infraspecies %>% select(contains("M", ignore.case = F))))


compo_taxa_meat_plant_ratio_cor =

compo_taxa_melt %>%
  filter(variable %in% id) %>%
  #filter(taxa == "family") %>%
  mutate(variable = variable %>% as.character) %>%
  arrange(variable) %>%
  merge(meat_plant_ratio %>% select(variable,meat_plant_ratio), by="variable") %>%
  group_by(tax_variable,taxa) %>%
  do(with(.,cor.test(.$value, .$meat_plant_ratio, method="spearman")) %>% broom::tidy()) %>%
  arrange(p.value)
  


compo_taxa_meat_plant_ratio_cor %>% filter(p.value < 0.05)


```


## large correlation with nutriscore

```{r, message=FALSE, warning=FALSE}

data("nutriscore_DI")




compo_taxa_nutriscore_cor =

compo_taxa_melt %>%
  filter(variable %in% id) %>%
  #filter(taxa == "family") %>%
  mutate(variable = variable %>% as.character) %>%
  arrange(variable) %>%
  merge(nutriscore_DI , by.x="variable", by.y="Patient_ID") %>%
  group_by(tax_variable,taxa) %>%
  do(with(.,cor.test(.$value, .$nutriscore_DI, method="spearman")) %>% broom::tidy()) %>%
  arrange(p.value)
  

compo_taxa_nutriscore_cor


```


```{r}


merge(compo_taxa_nutriscore_cor,compo_taxa_meat_plant_ratio_cor, by="tax_variable") %>%
  filter(abs(p.value.x)<0.01|abs(p.value.y)<0.01) %>% 
  ggplot() + geom_label(aes(estimate.x,estimate.y,label=tax_variable))




merge(
compo_taxa_food_unifrac_permanova %>%
  select(tax_variable,taxa,R2,`p.value`) %>%
  merge(compo_taxa_nutriscore_cor %>% select(tax_variable,taxa,estimate,`p.value`), by = c("tax_variable","taxa") ) %>%
  filter(p.value.x < 0.05, p.value.y < 0.05),
  
compo_taxa_food_unifrac_permanova %>%
  select(tax_variable,taxa,R2,`p.value`) %>%
  merge(compo_taxa_meat_plant_ratio_cor %>% select(tax_variable,taxa,estimate,`p.value`), by = c("tax_variable","taxa") ) %>%
  filter(p.value.x < 0.05, p.value.y < 0.05), by=c("tax_variable","taxa","R2","p.value.x"))




MSP_tax %>%
  filter(MSP_ID == "msp_0029")


merge(compo_taxa_meat_plant_ratio_cor %>% select(tax_variable,taxa,estimate,`p.value`),
compo_taxa_nutriscore_cor %>% select(tax_variable,taxa,estimate,`p.value`),
by = c("tax_variable","taxa")) %>%
  filter(p.value.x < 0.05, p.value.y < 0.05)
  



```



```{r}

#nutrient_taxa_cor

# infraspecies %>%
#   select(-contains("M", ignore.case = F), msp_name, msp_name_partition) %>%
#   select(msp_name_partition, msp_name, genus, phylum)


nutrient_taxa_cor %>%
  group_by(taxa,tax_variable) %>%
  filter(cor == max(abs(cor))) %>%
  arrange(desc(abs(cor))) %>%
  ggplot() + geom_density(aes(x=cor, color=taxa))




```




# tree based representation

```{r}
library(purrr)


compo_taxa_food_unifrac_permanova %>% 
  ungroup %>%
  mutate(tax_variable = ifelse(taxa=="genus", paste0("g_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="family", paste0("f_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="order", paste0("o_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="class", paste0("c_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="phylum", paste0("p_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="superkingdom", paste0("k_",tax_variable), tax_variable )) %>%
  filter(R2>0.02) %>% .$tax_variable -> tax_variable_select
  

infraspecies %>%
  mutate(root="root") %>%
  select(root,superkingdom,phylum,class,order,family,genus,msp_name, msp_name_partition) %>%
  filter(msp_name %in% species_select) %>%
  mutate(superkingdom=paste0("k_",superkingdom),
         phylum=paste0("p_",phylum), 
         class=paste0("c_",class),
         order=paste0("o_",order),
         family=paste0("f_",family), 
         genus=paste0("g_",genus)) %>%
  filter(
    msp_name_partition %in% tax_variable_select |
      msp_name %in% tax_variable_select |
      genus %in% tax_variable_select |
      family %in% tax_variable_select |
      class %in% tax_variable_select |
      order %in% tax_variable_select |
      phylum %in% tax_variable_select |
      superkingdom %in% tax_variable_select
    
    
  )   -> df




taxo_edges =

  map_df(2:ncol(df), ~select(df, (.x-1):.x) %>% setNames(c("source", "target")), .id="id") %>%
    group_by(id) %>%
    distinct() %>%
    ungroup() %>%
    mutate(id = colnames(df)[as.numeric(id)+1])



taxo_edges %>%
  mutate(unassigned =  grepl("unassigned", target)) -> taxo_edges
  
  
df


```


```{r, message=FALSE, warning=FALSE}



library(igraph)
library(ggraph)

gr <- graph_from_data_frame(cbind(taxo_edges[2:4],taxo_edges[1:3]))

names(V(gr)) %>% length



R2_permanova_diet =

compo_taxa_food_unifrac_permanova %>% 
  ungroup %>%
  mutate(tax_variable = ifelse(taxa=="genus", paste0("g_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="family", paste0("f_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="order", paste0("o_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="class", paste0("c_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="phylum", paste0("p_",tax_variable), tax_variable )) %>%
  mutate(tax_variable = ifelse(taxa=="superkingdom", paste0("k_",tax_variable), tax_variable )) %>%
  tibble::column_to_rownames("tax_variable") %>%
  .[names(V(gr)),] %>%
  .$R2
 

V(gr)$R2_permanova_diet = R2_permanova_diet

R2_boolean = R2_permanova_diet > 0.02

R2_boolean[1] = FALSE

V(gr)$R2_boolean = R2_boolean

V(gr)$unassigned = names(V(gr)) %>% grepl("unassigned",.)

V(gr)$label = names(V(gr))


dendro = ggraph(gr, layout = 'dendrogram', circular = TRUE) + 
    geom_edge_elbow(aes(filter = !unassigned)) +
    geom_node_point(aes(color = R2_permanova_diet, size=R2_permanova_diet, filter= !unassigned, alpha=R2_boolean)) +
    geom_node_text(aes(label=label, filter=!unassigned)) +
    scale_radius() +
    viridis::scale_color_viridis() +
    coord_fixed() #+ theme_void()

#dendro

dendro_g  <- ggplot_build(dendro)


dendro = ggraph(gr, layout = 'dendrogram', circular = TRUE) + 
    geom_edge_elbow(aes(filter = !unassigned), alpha=0.7) +
    geom_node_point(aes(alpha = R2_permanova_diet, size=R2_permanova_diet, filter= !unassigned)) +
    #geom_node_text(aes(label=label, filter=!unassigned)) +
    scale_radius() +
    #viridis::scale_color_viridis() +
    coord_fixed() + theme_void() + scale_alpha(expression(atop("\nTaxonomic tree nodes\nassociated with diet",paste("PERMANOVA R"^2))), labels = scales::percent) + scale_size(expression(atop("\nTaxonomic tree nodes\nassociated with diet",paste("PERMANOVA R"^2))), labels = scales::percent)

dendro=
dendro +
  new_scale("color") +
  geom_line(data=merge(dendro_g[[1]][[3]] , df, by.x="label", by.y="msp_name") %>%
              group_by(family.y) %>%
              mutate(n=n()) %>%
  #filter(order=="o_Clostridiales" & n>1) %>%
    filter(n>5) %>%
  arrange(x,y), aes(x=x,y=y, group=family.y %>% gsub("f_","",.), color=family.y %>% gsub("f_","",.)),alpha=0.5, size=5) + 
  scale_color_brewer("MSP taxonomic family", type="qual")

dendro
ggsave("Fig_R2_permanova_tree.pdf", h=7,w=12)



```


the MSP 220 (unclassified clostridiales, formely an eubacterium) can be splitted into two infraspecies.The MSP_220_cl1 is more associated with Diet. The difference is associated with Antibiotics biosynthesis, lipolysacharides metabolism and spore forming metabolism.

The MSP 70 (eubacterium rectale) can be splitted into two infraspecies. Infraspecies are more associated with diet (meat/plan ratio). cl1 have a flagelin module.


## R2 throughout lineage

```{r}
compo_taxa_food_unifrac_permanova %>%
  filter(taxa=="phylum")


```


when MSP can be splitted into subspecies, diet associations is always higher at subspecies

```{r}

infraspecies %>%
  select(superkingdom,phylum,class,order,family,genus, msp_name,msp_name_partition) %>%
  ungroup() %>%
  dplyr::rename(MSP=msp_name, Infraspecies=msp_name_partition) %>%
  tibble::rowid_to_column() %>%
  reshape2::melt(id.vars="rowid") %>%
  merge(compo_taxa_food_unifrac_permanova, by.y=c("tax_variable","taxa"), by.x=c("value","variable")) %>%
  filter(term=="value") %>%
  group_by(rowid) %>%
  mutate(n_level= n()) %>%
  filter(n_level > 6) %>%
  filter(!any((grepl("unassigned",value)))) %>%
  filter(any(R2>0.02)) %>%
  mutate(family_R2=ifelse(any(variable=="family"), R2[variable=="family"], 0)) %>%
  mutate(genus_R2=ifelse(any(variable=="genus"), R2[variable=="genus"], 0)) %>%
  mutate(MSP_R2=ifelse(any(variable=="MSP"), R2[variable=="MSP"], 0)) %>%
  mutate(Infraspecies_R2=ifelse(any(variable=="Infraspecies"), R2[variable=="Infraspecies"], 0)) %>%
  mutate(Associations = ifelse(genus_R2 >= Infraspecies_R2,"MSP or higher level","subspecies level")) %>%
  mutate(Associations = ifelse(MSP_R2 >= Infraspecies_R2,"MSP or higher level","subspecies level")) %>%
  filter(variable=="MSP") %>%
  group_by(value) %>%
  filter(Infraspecies_R2 == max(Infraspecies_R2)) %>%
  ungroup() %>%
  group_by(Associations) %>%
  summarise(n=n())
  

infraspecies %>%
  select(superkingdom,phylum,class,order,family,genus, msp_name,msp_name_partition) %>%
  ungroup() %>%
  dplyr::rename(MSP=msp_name, Infraspecies=msp_name_partition) %>%
  tibble::rowid_to_column() %>%
  reshape2::melt(id.vars="rowid") %>%
  merge(compo_taxa_food_unifrac_permanova, by.y=c("tax_variable","taxa"), by.x=c("value","variable")) %>%
  mutate(variable = variable %>% gsub("Infraspecies","subspecies",.)) %>%
  filter(term=="value") %>%
  group_by(rowid) %>%
  mutate(n_level= n()) %>%
  filter(n_level > 6) %>%
  filter(!any((grepl("unassigned",value)))) %>%
  filter(any(R2>0.02)) %>%
  mutate(family_R2=ifelse(any(variable=="family"), R2[variable=="family"], 0)) %>%
  mutate(genus_R2=ifelse(any(variable=="genus"), R2[variable=="genus"], 0)) %>%
  mutate(MSP_R2=ifelse(any(variable=="MSP"), R2[variable=="MSP"], 0)) %>%
  mutate(Infraspecies_R2=ifelse(any(variable=="subspecies"), R2[variable=="subspecies"], 0)) %>%
  mutate(Associations = ifelse(genus_R2 >= Infraspecies_R2,"MSP or higher level","subspecies level")) %>%
  mutate(Associations = ifelse(MSP_R2 >= Infraspecies_R2,"MSP or higher level","subspecies level")) %>%
  mutate(variable = factor(variable, levels = c("superkingdom","phylum","class","order","family","genus","MSP","subspecies"))) %>%
  ggplot() + geom_line(aes(x=variable,y=R2,group=rowid, col=Associations), alpha=0.5)  + 
  scale_color_manual("Higher diet\nassociation at",values = c("Red","Blue")) + 
  scale_y_continuous(labels = scales::percent, limits = c(0,0.035)) +
  xlab("Taxonomic levels") +
  ylab(expression(paste(R^2," (PERMANOVA)"))) +
  theme_minimal() -> p_R2_lineage


p_R2_lineage

ggsave("R2_lineage.pdf", h=4, w=7)

# which MSP where its respecetive infraspecies is more associated to the Diet?

infraspecies %>%
  select(superkingdom,phylum,class,order,family,genus, msp_name,msp_name_partition) %>%
  ungroup() %>%
  dplyr::rename(MSP=msp_name, Infraspecies=msp_name_partition) %>%
  tibble::rowid_to_column() %>%
  reshape2::melt(id.vars="rowid") %>%
  merge(compo_taxa_food_unifrac_permanova, by.y=c("tax_variable","taxa"), by.x=c("value","variable")) %>%
  filter(term=="value") %>%
  group_by(rowid) %>%
  mutate(n_level= n()) %>%
  filter(n_level > 6) %>%
  #filter(!any((grepl("unassigned",value)))) %>%
  #filter(any(R2>0.02)) %>%
  mutate(family_R2=ifelse(any(variable=="family"), R2[variable=="family"], 0)) %>%
  mutate(genus_R2=ifelse(any(variable=="genus"), R2[variable=="genus"], 0)) %>%
  mutate(MSP_R2=ifelse(any(variable=="MSP"), R2[variable=="MSP"], 0)) %>%
  mutate(Infraspecies_R2=ifelse(any(variable=="Infraspecies"), R2[variable=="Infraspecies"], 0)) %>%
   mutate(Associations = ifelse(genus_R2 >= Infraspecies_R2,"MSP or higher level","Infraspecies level")) %>%
  mutate(Associations = ifelse(MSP_R2 >= Infraspecies_R2,"MSP or higher level","Infraspecies level")) %>%
  filter(Associations == "Infraspecies level", variable=="MSP") %>%
  ungroup() %>%
  select(value) %>% unique -> MSP_keep_infraspecies
 
save(MSP_keep_infraspecies, file="MSP_keep_infraspecies.rda")

#131 MSP where one of its infraspecies is better associated to diet, among them how many R2>2% ?




MSP_tax %>%
  filter(MSP_ID=="msp_0038")



```


```{r, fig.height=7, fig.width=7}


cowplot::plot_grid(dendro,p_R2_lineage, nrow=2, rel_heights = c(2,1))



```



## read mass vs permanova


```{r}

rbind(infraspecies %>%
    mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
    filter(msp_name %in% species_select) %>%
    filter(msp_name %in% MSP_keep_infraspecies$value) %>%
    select(msp_name_partition,contains("M", ignore.case = F) ) %>%
    merge(compo_taxa_food_unifrac_permanova %>% select(tax_variable,R2),., by.y="msp_name_partition", by.x="tax_variable"),
  


  infraspecies %>%
    mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
    filter(msp_name %in% species_select) %>%
    filter(!(msp_name %in% MSP_keep_infraspecies$value)) %>%
    select(msp_name,contains("M", ignore.case = F) ) %>%
    group_by(msp_name) %>%
    summarize_at(vars(contains("M", ignore.case = F)),function(x){sum(x)}) %>%
    ungroup() %>%
    merge(compo_taxa_food_unifrac_permanova %>% select(tax_variable,R2),., by.y="msp_name", by.x="tax_variable")) %>%
    
  select(-taxa) %>%
  mutate(R2_class = cut(R2, c(0,0.01,0.02,1), include.lowest = TRUE)) %>%
  select(-R2) %>%
  ungroup() %>%
  reshape2::melt(id.vars=c("tax_variable","R2_class")) %>%
  group_by(R2_class, variable) %>%
  summarize(read_mass=sum(value)) %>%
  ungroup() %>%
  #mutate(variable = forcats::fct_reorder(variable, as.numeric(R2_class), fun = mean)) %>%
  #arrange(R2_class) %>%
  merge(.,cazotypes, by.x="variable",by.y="row.names") %>%
  ggplot(aes(x=forcats::fct_reorder2(factor(variable), R2_class, read_mass, function(x, y) {sum(y[x == "(0.02,1]"])}), 
               y=read_mass, fill= factor(R2_class))) + 
  geom_bar( stat="identity", color=NA, position="fill", width=1) + 
  scale_y_continuous(labels = scales::percent) + theme_minimal() +
  scale_fill_manual("R2",labels = c("<1%","1-2%",">2%"), values=c("red","grey","blue")) + facet_wrap(~cazotype, scales="free") +
  ylab("relative read mass") + xlab("subjects") + 
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())


```

```{r, fig.height=4, fig.width=5}


R2_permanova_diet_cazy_read_mass_df =  
rbind(
  infraspecies %>%
    mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
    filter(msp_name %in% species_select) %>%
    #filter(msp_name %in% MSP_keep_infraspecies$value) %>%
    select(msp_name_partition,contains("M", ignore.case = F) ) %>%
    merge(compo_taxa_food_unifrac_permanova %>% select(tax_variable,R2),., by.y="msp_name_partition", by.x="tax_variable"),
  


  infraspecies %>%
    mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
    filter(msp_name %in% species_select) %>%
    #filter(!(msp_name %in% MSP_keep_infraspecies$value)) %>%
    select(msp_name,contains("M", ignore.case = F) ) %>%
    group_by(msp_name) %>%
    summarize_at(vars(contains("M", ignore.case = F)),function(x){sum(x)}) %>%
    ungroup() %>%
    merge(compo_taxa_food_unifrac_permanova %>% select(tax_variable,R2),., by.y="msp_name", by.x="tax_variable")
  ) %>%
    
  #select(-taxa) %>%
  mutate(R2_class = cut(R2, c(0,0.01,0.02,1), include.lowest = TRUE)) %>%
  select(-R2) %>%
  ungroup() %>%
  reshape2::melt(id.vars=c("tax_variable","R2_class","taxa")) %>%
  group_by(R2_class, variable,taxa) %>%
  summarize(read_mass=sum(value)) %>%
  ungroup() %>%
  #mutate(variable = forcats::fct_reorder(variable, as.numeric(R2_class), fun = mean)) %>%
  #arrange(R2_class) %>%
  merge(.,cazotypes, by.x="variable",by.y="row.names") %>%
  merge(.,enterotypes %>% filter(Visit=="V4") %>% select(Patient_ID,enterotypes), by.x="variable",by.y="Patient_ID") 


R2_permanova_diet_cazy_read_mass_df %>%
  filter(R2_class=="(0.02,1]") %>%
ggplot() + geom_boxplot(aes(x=cazotype %>% as.character,y=read_mass,fill=taxa %>% as.character)) + 
  scale_y_log10(label=scales::percent) + 
  scale_fill_brewer("taxa level",type="qual") + ylab("Relative microbial read mass\nassociated with Diet\n(R2 > 2%)") +
  theme_classic() + annotate("text", x=3,y=0.025, label="p < 0.05", size=3) +
  xlab("Cazotypes") -> p_cazotypes_read_mass

p_cazotypes_read_mass

R2_permanova_diet_cazy_read_mass_df %>% 
  filter(taxa=="MSP") %>%
  filter(R2_class=="(0.02,1]") %>%
  with(.,pairwise.wilcox.test(read_mass , cazotype %>% as.character, data=., p.adjust.method="fdr"))



R2_permanova_diet_cazy_read_mass_df %>%
  filter(R2_class=="(0.02,1]") %>%
ggplot() + geom_boxplot(aes(x=enterotypes %>% as.character,y=read_mass,fill=taxa %>% as.character)) + 
  scale_y_log10(label=scales::percent) + 
  scale_fill_brewer("taxa level",type="qual") + ylab("Relative microbial read mass\nassociated with Diet\n(R2 > 2%)") +
  theme_classic() + annotate("text", x=3,y=0.025, label="p < 0.05", size=3) +
  xlab("Enterotypes") -> p_enterotypes_read_mass

p_enterotypes_read_mass


# R2_permanova_diet_cazy_read_mass_df =   
#   infraspecies %>%
#   mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
#   #filter(msp_name_partition %in% infraspecies_select) %>%
#   filter(msp_name %in% species_select) %>%
#   select(msp_name,contains("M", ignore.case = F) ) %>%
#   merge(compo_taxa_food_unifrac_permanova %>% select(tax_variable,R2), by.x="msp_name", by.y="tax_variable") %>%
#   select(-taxa) %>%
#   mutate(R2_class = cut(R2, c(0,0.01,0.02,1), include.lowest = TRUE)) %>%
#   select(-R2) %>%
#   ungroup() %>%
#   reshape2::melt(id.vars=c("msp_name","R2_class")) %>%
#   group_by(R2_class, variable) %>%
#   summarize(read_mass=sum(value)) %>%
#   ungroup() %>%
#   #mutate(variable = forcats::fct_reorder(variable, as.numeric(R2_class), fun = mean)) %>%
#   #arrange(R2_class) %>%
#   merge(.,cazotypes, by.x="variable",by.y="row.names") %>%
#   merge(.,enterotypes %>% filter(Visit=="V4") %>% select(Patient_ID,enterotypes), by.x="variable",by.y="Patient_ID") 
# 
# R2_permanova_diet_cazy_read_mass_df %>%
#   filter(R2_class=="(0.02,1]") %>%
# ggplot() + geom_boxplot(aes(x=cazotype,y=read_mass,fill=cazotype %>% as.character)) + 
#   scale_y_log10(label=scales::percent) + 
#   scale_fill_brewer("Cazotypes",type="qual") + ylab("Relative microbial read mass\nassociated with Diet\n(R2 > 2%)") +
#   theme_classic() + guides(fill=FALSE) + annotate("text", x=3,y=0.1, label="p < 0.05", size=3)
# 
# 
# R2_permanova_diet_cazy_read_mass_df %>% 
#   filter(R2_class=="(0.02,1]") %>%
#   with(.,pairwise.wilcox.test(read_mass , cazotype %>% as.character, data=., p.adjust.method="fdr"))



```


## Eubacterium rectale infraspecies



```{r}

infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(species == "Eubacterium rectale") %>%
  filter(msp_name_partition != "msp_0070_unassigned") %>%
  select(msp_name_partition, contains("M", ignore.case = FALSE)) %>%
  reshape2::melt(id.vars="msp_name_partition") %>%
  #group_by(variable) %>%
  #filter(value==max(value)) %>%
  filter(msp_name_partition=="msp_0070_cl_1") %>%
  merge(meat_plant_ratio %>% select(variable,meat_plant_ratio, A1, `Plant-based`,`Animal-based`),by="variable") %>%
  #filter(value>0) %>%
  #with(.,cor.test(.$`Plant-based`,.$value+10^-4, method="spearman"))
  with(.,cor.test(.$`meat_plant_ratio`,.$value+10^-4, method="spearman"))



infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(species == "Eubacterium rectale") %>%
  filter(msp_name_partition != "msp_0070_unassigned") %>%
  select(msp_name_partition, contains("M", ignore.case = FALSE)) %>%
  reshape2::melt(id.vars="msp_name_partition") %>%
  #group_by(variable) %>%
  #filter(value==max(value)) %>%
  filter(msp_name_partition=="msp_0070_cl_1") %>%
  merge(meat_plant_ratio , by="variable") %>%
  dplyr::rename(patient="variable", erectale="value") %>%
  reshape2::melt(id.vars=c("patient","msp_name_partition","erectale")) %>%
  group_by(variable) %>%
  do(with(.,cor.test(.$value, .$erectale, method="spearman")) %>% broom::tidy()) %>%
  arrange(p.value)
  
infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(species == "Eubacterium rectale") %>%
  filter(msp_name_partition != "msp_0070_unassigned") %>%
  select(msp_name_partition, contains("M", ignore.case = FALSE)) %>%
  reshape2::melt(id.vars="msp_name_partition") %>%
  #group_by(variable) %>%
  #filter(value==max(value)) %>%
  #filter(value>0) %>%
  filter(msp_name_partition=="msp_0070_cl_2") %>%
  merge(nutrients_data , by.x="variable", by.y="ID") %>%
  dplyr::rename(patient="variable", erectale="value") %>%
  reshape2::melt(id.vars=c("patient","msp_name_partition","erectale")) %>%
  group_by(variable) %>%
  do(with(.,cor.test(.$value, .$erectale, method="spearman")) %>% broom::tidy()) %>%
  arrange(p.value)
  

infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(species == "Eubacterium rectale") %>%
  filter(msp_name_partition != "msp_0070_unassigned") %>%
  select(msp_name_partition, contains("M", ignore.case = FALSE)) %>%
  reshape2::melt(id.vars="msp_name_partition") %>%
  #group_by(variable) %>%
  #filter(value==max(value)) %>%
  #filter(msp_name_partition=="msp_0070_cl_1") %>%
  merge(nutrients_data , by.x="variable", by.y="ID") %>%
  #filter(value>0) %>%
  #with(.,wilcox.test(.$`Plant-based`~.$msp_name_partition))
  #with(.,cor.test(.$`Plant-based`,.$value, method="spearman"))
  #ggplot() + geom_boxplot(aes(x=msp_name_partition,y=meat_plant_ratio,col=msp_name_partition)) 
  #ggplot() + geom_boxplot(aes(x=msp_name_partition,y=`Plant-based`)) 
  filter(msp_name_partition=="msp_0070_cl_1") %>%
  ggplot() + 
  geom_point(aes(x=value+10^-4,y=`Vit B12 ug`,col=msp_name_partition)) + 
  #geom_smooth(aes(x=value+10^-4,y=meat_plant_ratio), method = "lm") +
  annotate("text",y=-2,x=0.003, label="Rho=0.31, p<0.05") +
  scale_x_log10(label=scales::percent) + 
  ylab("Vitamin B12 (ug)") +
  xlab("Eubacterium rectale infraspecies\nwith flagelin encoded genes") -> p_Erectale_VitB12






infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(species == "Eubacterium rectale") %>%
  filter(msp_name_partition != "msp_0070_unassigned") %>%
  select(msp_name_partition, contains("M", ignore.case = FALSE)) %>%
  reshape2::melt(id.vars="msp_name_partition") %>%
  #group_by(variable) %>%
  #filter(value==max(value)) %>%
  #filter(msp_name_partition=="msp_0070_cl_1") %>%
  merge(meat_plant_ratio %>% select(variable,meat_plant_ratio, A1, `Plant-based`,`Animal-based`),by="variable") %>%
  dplyr::rename(value_erectale="value") %>%
  group_by(msp_name_partition) %>%
  select(meat_plant_ratio,value_erectale) %>%
  reshape2::melt(id.vars=c("msp_name_partition","value_erectale")) %>%
  mutate(detected=ifelse(value_erectale>0,"yes","no")) %>%
  ungroup() %>%
  #select(-value) %>%
  #ggplot() + geom_boxplot(aes(x=msp_name_partition,y=value,fill=detected))
  filter(msp_name_partition=="msp_0070_cl_1") %>%
  #filter(detected=="yes") %>%
  with(., wilcox.test(.$value~ .$detected))


infraspecies %>%
  mutate_at(vars(contains("M", ignore.case = F)),function(x){x/sum(x)}) %>%
  #filter(msp_name_partition %in% infraspecies_select) %>%
  filter(species == "Eubacterium rectale") %>%
  filter(msp_name_partition != "msp_0070_unassigned") %>%
  select(msp_name_partition, contains("M", ignore.case = FALSE)) %>%
  reshape2::melt(id.vars="msp_name_partition") %>%
  #group_by(variable) %>%
  #filter(value==max(value)) %>%
  #filter(msp_name_partition=="msp_0070_cl_1") %>%
  merge(meat_plant_ratio %>% select(variable,meat_plant_ratio, A1, `Plant-based`),by="variable") %>%
  filter(value>0) %>%
  #with(.,wilcox.test(.$`Plant-based`~.$msp_name_partition))
  #with(.,cor.test(.$`Plant-based`,.$value, method="spearman"))
  #ggplot() + geom_boxplot(aes(x=msp_name_partition,y=meat_plant_ratio,col=msp_name_partition)) 
  #ggplot() + geom_boxplot(aes(x=msp_name_partition,y=`Plant-based`)) 
  ggplot() + 
  geom_point(aes(x=value+10^-4,y=meat_plant_ratio,col=msp_name_partition)) + 
  #geom_smooth(aes(x=value+10^-4,y=meat_plant_ratio), method = "lm") +
  annotate("text",y=-2,x=0.003, label="Rho=0.23, p<0.05") +
  scale_x_log10(label=scales::percent) + 
  ylab("Meat\\Plant based ratio (log2)") +
  xlab("Eubacterium rectale infraspecies\nwith flagelin encoded genes") -> p_Erectale_meat_plant_plot
  


p_Erectale_meat_plant_plot

ggsave("p_Erectale_meat_plant_plot.pdf")


```


```{r}


load("Erectale.rda")

#Erectale$clusters_msp

#plot_MSP(binary_msp = Erectale$binary_msp, clusters_msp =  Erectale$clusters_msp$sample)

genes_hclust = 
  Erectale$binary_msp %>%
            filter(module_name  != "core") %>%
            select(-module_name,-genes.id) %>%
            as.matrix %>% .[,order(Erectale$clusters_msp$sample$partition)] %>% dist(method="binary") %>% hclust(method="centroid")




Erectale$binary_msp %>% 
  filter(module_name  != "core") %>% 
  select(-module_name,-genes.id) %>% 
  as.matrix %>% .[genes_hclust$order,order(Erectale$clusters_msp$sample$partition)] %>%
  as.data.frame() %>%
  tibble::rowid_to_column() %>%
  reshape2::melt(id.vars="rowid") %>%
  merge(Erectale$clusters_msp$sample$partition %>%
   as.matrix() %>% as.data.frame, by.x="variable", by.y="row.names") %>%
  dplyr::rename(Infraspecies = V1) %>%
  group_by(rowid) %>%
  mutate(prev=sum(value)/n()) %>%
  filter(prev>0.25) %>%
  ggplot() + geom_tile(aes(x=variable,y=rowid %>% as.character, fill=value %>% as.character)) + 
  scale_fill_manual("", labels=c("Undetected","Detected"), values=c("dark grey","yellow")) +
  facet_wrap(~Infraspecies, scales = "free_x") +
  theme_bw() +
  theme(axis.ticks=element_blank()) +
  theme(axis.text=element_blank()) +
  ylab("Genes") +
  xlab(substitute(paste("Metagenomic ", italic('E. rectale'), " subspecies"))) -> p_erectale_infraspecies

  
  
save(p_erectale_infraspecies, file="p_erectale_infraspecies.rda" )

p_erectale_infraspecies

ggsave("p_erectale_infraspecies.pdf")


```

extract flagelin genes
```{r}

og=vroom::vroom("../data-raw/metahit.og.csv",delim = ";")

og %>% select(-1) %>%
  merge(Erectale$test_msp, by="genes.id") %>% write.csv(file = "erectale_cog_significant_genes.csv")

system.file("data-raw","erectale_flagelin.csv", package="IBSFood")


flagelin_genes = read.csv2("../data-raw/erectale_flagelin.csv")



```


```{r}

Erectale$binary_msp %>% 
  filter(module_name  != "core") %>% 
  select(-module_name) %>%
  tibble::column_to_rownames("genes.id") %>%
  as.matrix %>% .[genes_hclust$order,order(Erectale$clusters_msp$sample$partition)] %>%
  as.data.frame() %>%
  tibble::rownames_to_column("genes.id") %>%
  filter(genes.id %in% flagelin_genes$genes_id) %>%
  as.data.frame() %>%
  #tibble::rowid_to_column() %>%
  reshape2::melt(id.vars="genes.id") %>%
  merge(Erectale$clusters_msp$sample$partition %>%
   as.matrix() %>% as.data.frame, by.x="variable", by.y="row.names") %>%
  dplyr::rename(Infraspecies = V1) %>%
  group_by(genes.id) %>%
  mutate(prev=sum(value)/n()) %>%
  filter(prev>0.25) %>%
  ggplot() + geom_tile(aes(x=variable,y=genes.id %>% as.character, fill=value %>% as.character)) + 
  scale_fill_manual("", labels=c("Undetected","Detected"), values=c("dark grey","yellow")) +
  facet_wrap(~Infraspecies, scales = "free_x") +
  theme_bw() +
  theme(axis.ticks=element_blank()) +
  theme(axis.text=element_blank()) +
  ylab("Flagelin encoded genes") +
  xlab(substitute(paste("Metagenomic ", italic('E. rectale'), " subspecies"))) -> p_erectale_infraspecies_flagelin

```



## final plot (Figure 2)
```{r, fig.height=7, fig.width=12}

# cowplot::plot_grid(
#   cowplot::plot_grid(dendro,p_R2_lineage, nrow=2, rel_heights = c(2,1), labels = c("A","C")),
#   cowplot::plot_grid(p_cazotypes_read_mass,p_erectale_infraspecies,nrow=2, labels = c("B","D")), 
#   ncol=2, rel_widths = c(2,1))

cowplot::plot_grid(
  cowplot::plot_grid(dendro,p_R2_lineage, nrow=2, rel_heights = c(2,1), labels = c("a","b")),
  cowplot::plot_grid(p_erectale_infraspecies,p_erectale_infraspecies_flagelin,nrow=2, labels = c("c",""), rel_heights = c(2,1)), 
  ncol=2, rel_widths = c(2,1))

ggsave("Figure2_infraspecies_diet_flagelin_small.pdf")


```





## import cazy MSP

```{r}

MSP_CAZY_df = read.csv2("MSP_CAZY_df.csv", row.names=1)




MSP_CAZY_df %>%
  filter(msp_name %in% species_select) %>%
  group_by(msp_name, cazy.family) %>%
  summarize(n=n()) %>%
  filter(cazy.family %in%  c("GH2","GH13","GH43","GH20","GH92","CBM32","CBM26","GT2") ) %>%
  reshape2::dcast(msp_name~cazy.family, value.var="n", fill=0) %>%   
  merge(.,compo_taxa_food_unifrac_permanova, by.x="msp_name", by.y="tax_variable") %>% 
  filter(term=="value") %>%
  ggplot() + geom_boxplot(aes(y=R2,x=!(GH2>0&GH13>0)))


MSP_CAZY_df %>%
  filter(msp_name %in% species_select) %>%
  group_by(msp_name, cazy.family) %>%
  summarize(n=n()) %>%
  filter(cazy.family %in%  c("GH2","GH13","GH43","GH20","GH92","CBM32","CBM26","GT2") ) %>%
  reshape2::dcast(msp_name~cazy.family, value.var="n", fill=0) %>%   
  merge(.,compo_taxa_food_unifrac_permanova, by.x="msp_name", by.y="tax_variable") %>% 
  filter(term=="value") %>%
  mutate(test=!(GH2>0|GH13>0|GT2>0)) %>%
  wilcox.test(R2~test, data=.)



MSP_CAZY_df %>%
  filter(msp_name %in% species_select) %>%
  group_by(msp_name, cazy.family) %>%
  summarize(n=n()) %>%
  mutate(n=ifelse(n>0,1,0)) %>%
  filter(cazy.family %in%  c("GH2","GH13","GH43","GH20","GH92","CBM32","CBM26","GT2") ) %>%
  reshape2::dcast(msp_name~cazy.family, value.var="n", fill=0) %>%   
  merge(.,compo_taxa_food_unifrac_permanova, by.x="msp_name", by.y="tax_variable") -> MSP_CAZY_df_casted


MSP_CAZY_df_casted %>%
  mutate(test=GH13+GH2+GT2+GH20+GH92+GH43+CBM32+CBM26+GH92) %>%
ggplot() + geom_boxplot(aes(x=test%>% as.character,y=R2))



MSP_CAZY_df_casted %>%
  #filter(R2 > 0.02 | R2 < 0.005) %>%
  mutate(R2_scale = scale(R2)) %>%
  with(.,lm(R2_scale~GH13*GH2*GH43*GH20*GH92*GT2)) %>% summary()








```



```{r}


MSP_CAZY_df = read.csv2("MSP_CAZY_df.csv", row.names=1)



MSP_CAZY_df %>%
  filter(msp_name %in% species_select) %>%
  group_by(msp_name, cazy.family) %>%
  summarize(n=n()) %>%
  mutate(n=ifelse(n>0,1,0)) %>%
  #filter(cazy.family %in%  c("GH2","GH13","GH43","GH20","GH92","CBM32","CBM26","GT2") ) %>%
  #reshape2::dcast(msp_name~cazy.family, value.var="n", fill=0) %>%   
  ungroup() %>%
  group_by(msp_name) %>%
  summarise(n=sum(n)) %>%
  merge(.,compo_taxa_food_unifrac_permanova, by.x="msp_name", by.y="tax_variable") %>%
  with(., cor.test(.$n , .$R2, data=., method="spearman") %>% broom::tidy())
  
  


MSP_CAZY_df %>%
  filter(msp_name %in% species_select) %>%
  group_by(msp_name, cazy.family) %>%
  summarize(n=n()) %>%
  mutate(n=ifelse(n>0,1,0)) %>%
  #filter(cazy.family %in%  c("GH2","GH13","GH43","GH20","GH92","CBM32","CBM26","GT2") ) %>%
  #filter(grepl("GH|PL|GT",cazy.family)   ) %>%
  #reshape2::dcast(msp_name~cazy.family, value.var="n", fill=0) %>%   
  ungroup() %>%
  group_by(msp_name) %>%
  summarise(n=sum(n)) %>%
  merge(.,compo_taxa_food_unifrac_permanova, by.x="msp_name", by.y="tax_variable") %>%
  filter(term=="value") %>% 
  mutate(n_class=cut(n,seq(0,150,30), include.lowest = TRUE), R2_class=cut(R2,c(0,0.01,0.02,0.03), include.lowest = TRUE)) %>%
  ggplot() + geom_bar(aes(fill=R2_class,x=n_class), position="fill") + viridis::scale_fill_viridis(discrete = TRUE)







```







